{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# A computational framework for the efficient formulation and calibration of large-scale, thermodynamically consistent kinetic models and its application"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Abstract"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true,
    "pycharm": {
     "name": "#%% \n"
    }
   },
   "source": [
    "## Introduction"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Cells use signal transduction to respond to extracellular cues. Extracellular ligands bind to RTKs that initiate phosphorylation cascades. Depending on protein abundance, different protein complexes will be assembled, which determines which part of downstream signaling will be activated. In contrast to phosphorylation, assembly of protein complexes does not consume energy. Consumption of energy can be checked via wegscheider cycle conditions. Enumerating all cycles is an NP-hard problem, so structured approach is necessary to ensure energy conservation. Faeder et al have introduced thermodynamic rule based modelling which derives kinetic rates from the specified gibbs free energies. Binding reactions generated using thermodynamic framework inherently satisfy energy conservation, making them biophysically more plausible. Also allows specification of rules that consume energy, such as phosphorylation or dephosphorylation by explicitely specifying consumption of cellular energy sources such as ATP or NADPH.\n",
    "Derivation of kinetic rates from gibbs free energies are carried out using arrhenius equation, but more modern formulations exist. Formulation provides an intuition why log-scale is more natural scale for most parameters and why optimization etc works better when log-transformed."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Energy conservation and corresponding wegscheider conditions introduce an constraint on parameters, which may suggest that this yields smaller less complex models. However, thermodynamically sounds models require reversibility of all reactions, which usually leads to more complex models. Specifically, binarization of reactions usually leads to non-reversible reactions that may violate energy conservation. An example for binarization of a reaction would be binding of RAS. RAF has a much higher affinity to gtp loaded Ras (RAS-gtp) compared to gdp loaded Ras (RAS-gdp), so it is tempting to implement a conditional reaction where RAF only binds RAS-gtp, but not RAS-gdp. As RAS is a GTPase, it can hydrolyse the loaded gtp to gdp, converting RAS-gtp to RAS-gdp. This reaction creates a complex in which RAS-gdp is bound to RAF. As the corresponding binding reaction was explicitely not created, the unbinding reaction for RAS-gdp and RAF would necessarily be irreversible and may thus violate energy conservation. Implementing a non-binarized reaction where RAF binds both RAS-gtp and RAS-gdp can easily be done in the thermodynamic framework, where the difference in affinity would be implemented as difference in Gibbs Free Energy, and would be guaranteed to satisfy wegscheider cycle conditions and thus satisfy energy conservation. However, the non-binarized, thermodynamic model needs to account for more species, which may pose computational challenges when the model describes more complex oligomerization processes. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Tools such as NFSim are particularly suited to simulate "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Formulating a Thermodynamic Model of RAF inhibition in PySB"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First import various classes that are necessary to define rules and initialize the model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pysb import Model, Compartment, Rule, EnergyPattern, Expression\n",
    "model = Model('thermo_raf');"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Protein Species\n",
    "As first step of the model construction, we have to define all molecules which we want to account for. In PySB, this is done through using the Monomer class. Each monomer is initialized by a unique identifier, a list of site as well as, optionally, a list of states for each site. Sites without states are typically describe the bonds involved in complex formation, while sites with states can also describe (post-translational) protein modifitations such as mutations or phosphorylation. \n",
    "\n",
    "Here, we we will define monomers _BRAF_ and _CRAF_ which both feature the interaction sites _RBD_ (interaction domain with RAS), _mek_ (interaction domain with MEK), _raf_ (dimerization domain) and _rafi_ (inhibitor binding site). For BRAF we also add a mutation sites for the 600th amino acid, which can take values _V_ (valine, wildtype) or _E_ (glutamic acid, oncogenic mutation)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pysb import Monomer\n",
    "Monomer('BRAF', ['AA600', 'RBD', 'mek', 'raf', 'rafi'], {'AA600': ['E', 'V']})\n",
    "Monomer('CRAF', ['RBD', 'mek', 'raf', 'rafi']);"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, we will specify initial concentrations. For this model, the concentration of all molecular species will be in $\\mu\\mathrm{M}$. However, we will use protein counts per cell derived from absolute proteomics to inform initial conditions. Accordingly, parameters that define initial concentrations have to be transformed from molecule per cell to $\\mu\\mathrm{M}$, which is achieved by dividing them by cell volume (here assumed to be $1pL = 10^{-12} L$) and Avogadro constant ($6.022 10^{23}$ molecules) and multiplying with $10^{6}$ to account for the unit prefix $\\mu$.\n",
    "\n",
    "In the following code we first introduce two expressions that define the avogadro constant and cell volume. PySB automatically creates workspace variables with the name of the expressions, which is passed as first argument, that can then later be used as argument to functions or as part of more complex expressions. Next, we introduce two parameter _BRAF_0_ and _CRAF_0_ that define initial abundances as molecules per cell and convert these abundances to concentrations in $\\mu\\mathrm{M}$ using the expressions _initBRAF_ and _initCRAF_. These expressions are then used to define the initial abundances of two molecular species for CRAF and BRAF. A molecular species is defined by a pattern, which is created by invoking the respective monomer with the state of each site as keyword argument. For initial conditions, the respective patterns have to be explicit, i.e., all states have to defined. Here, we specify that all sites are unbound (denoted by _None_), with the exception of the the _AA600_ site, for which we have to pick ony of the previously specified states, where we pick the oncogenic variant denoted by _E_."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pysb import Parameter, Expression, Initial\n",
    "import sympy as sp\n",
    "\n",
    "# define Avogadro constant and volume as hardcoded expressions\n",
    "Expression('N_Avogadro', sp.Float(6.02214076e+23))\n",
    "Expression('volume', sp.Float(1e-12))\n",
    "\n",
    "# define initial abundance parameters \n",
    "Parameter('BRAF_0', 0.0)\n",
    "Parameter('CRAF_0', 0.0)\n",
    "\n",
    "# convert initial abundances to concentrations\n",
    "Expression('initBRAF', 1000000.0*BRAF_0/(N_Avogadro*volume))\n",
    "Expression('initCRAF', 1000000.0*CRAF_0/(N_Avogadro*volume))\n",
    "\n",
    "# define initial molecular species\n",
    "Initial(BRAF(AA600='E', RBD=None, mek=None, raf=None, rafi=None), initBRAF)\n",
    "Initial(CRAF(RBD=None, mek=None, raf=None, rafi=None), initCRAF);"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Protein Interactions\n",
    "In rule-based modeling, all dynamic interactions are specified as rules. Rules are generalizations of biochemical reactions and their action is defined by a reactant pattern and a product pattern. When applied to a molecular species, the action of the rule, i.e., the respective biochemical reaction, is implemented by replacing one occurence of the reactant pattern by the product pattern in a molecular species. The reactant pattern and product pattern do not have to be explicit and a single rule can define multiple different biochemical reactions, depending in how often and in how many molecular species the reactant pattern occurs.\n",
    "\n",
    "Non-thermodynamic rules additionally require forward and, if applicable, reverse reaction rates. In contrast, thermodynamic rules are specified in terms of activation energy and the phenomenological constant $\\phi$, which encodes whether changes in free energy affect forwards or reverse rates. For thermodynamic rules, the reaction rate also depends on EnergyPatterns that may apply to educts or products of the reaction. This means that all reactions generated by non-thermodynamic rules have the same reaction rates, unless the rate is a local function. In contrast, reactions generated by thermodynamic rules can have context dependent reaction rates.\n",
    "\n",
    "In the following we will specify the dimerization of RAF molecules. We introduce three parameters, the forward rate $k_f$, the dissociation constant $k_D$ and the thermodynamic parameter $\\phi$. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "Parameter('bind_RAF_RAF_Ea', 10.0)\n",
    "Parameter('bind_RAF_RAF_dG', 0.01)\n",
    "Parameter('bind_RAF_RAF_phi', 1.0)\n",
    "\n",
    "Expression('Ea0_bind_RAF_RAF', -bind_RAF_RAF_phi*bind_RAF_RAF_dG - bind_RAF_RAF_Ea);"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To construct homo- and heterodimerization rules for all RAF paralogs, we use itertools to loop over all combinations of BRAF and CRAF. This assumes equal affinities and dynamics for all proteins."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "import itertools as itt\n",
    "\n",
    "for RAF1, RAF2 in itt.combinations_with_replacement([BRAF, CRAF], 2):\n",
    "    Rule(f'{RAF1.name}_and_{RAF2.name}_bind_and_dissociate', \n",
    "         RAF1(raf=None) + RAF2(raf=None) | RAF1(raf=1) % RAF2(raf=1), \n",
    "         bind_RAF_RAF_phi, Ea0_bind_RAF_RAF, energy=True)\n",
    "    EnergyPattern(f'ep_bind_{RAF1.name}_{RAF2.name}', \n",
    "                  RAF1(raf=1) % RAF2(raf=1), bind_RAF_RAF_dG);"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### RAF inhibitor\n",
    "Here we introduce a RAF inhibitor RAFi. We assume that the inhibitor is added to the cell medium at some point and quickly diffuses in and out of the cell. As extracellular space is much bigger than the volume of a cell, we can assume an infinite reservoir of molecules on the outside and assume that the intracellular inhibitors concentration will be unaffected by intracellular reactions. Accordingly, we specify the respective initial as constant, which means that the respective molecular species can participate in reactions, but it's concentration will remain constant."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "Monomer('RAFi', ['raf'])\n",
    "\n",
    "Initial(RAFi(raf=None), Parameter('RAFi_0', 0.0), fixed=True);"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next we define binding reactions for RAFi with both BRAF and CRAF, again assuming the same affinities and activation energies for both reactions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "Parameter('bind_RAFi_RAF_Ea', 10.0)\n",
    "Parameter('bind_RAFi_RAF_dG', 0.01)\n",
    "Parameter('bind_RAFi_RAF_phi', 1.0)\n",
    "Expression('Ea0_bind_RAFi_RAF', -bind_RAFi_RAF_phi*bind_RAFi_RAF_dG - bind_RAFi_RAF_Ea)\n",
    "\n",
    "for RAF in [BRAF, CRAF]:\n",
    "    Rule(f'RAFi_and_{RAF.name}_bind_and_dissociate', \n",
    "         RAFi(raf=None) + RAF(rafi=None) | RAFi(raf=1) % RAF(rafi=1), \n",
    "         bind_RAFi_RAF_phi, Ea0_bind_RAFi_RAF, energy=True)\n",
    "    EnergyPattern(f'ep_bind_{RAF.name}_RAFi', RAFi(raf=1) % RAF(rafi=1), bind_RAFi_RAF_dG);"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Paradoxical Activation\n",
    "RAF inhibitors inhibit growth for BRAF mutants but promote growth for NRAS mutants. At the strucural level, this can be rationalized by assuming that RAF inhibitors have higher affinity towards drug-unbound RAF dimers and lower affinity towards drug bound RAF dimers. The symmetry conveyed by the no energy consuming nature of molecular binding reactions, implies that RAF inhibitors, at low to medium concentrations, promote dimerization but incompletely inhibit signaling as they only bind to one of the two protomers in a dimer. As MAPK signaling in NRAS mutant cells is mediated by RAF dimers, respective signaling is amplified, leading to increased growth. In the thermodynamic framework, this can be implemented by adding an energy pattern that controls the Gibbs free energy of RAF-RAF-RAFi trimers. Note that we do not specify how these trimers are formed, so the change in energy will equally apply to the rates of all reactions that produce these trimers. In this example, an decrease in energy would equally increase RAF dimerization and inhbitor binding, thus implementing the previously described symmetry."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "Parameter('ep_RAF_RAF_mod_RAFi_single_ddG', 0.001)\n",
    "for RAF1, RAF2 in itt.product([BRAF, CRAF], repeat=2):\n",
    "    EnergyPattern(f'ep_{RAF1.name}_{RAF2.name}_mod_RAFi_single', \n",
    "                  RAF1(raf=1, rafi=None) % RAF2(raf=1, rafi=2) % RAFi(raf=2), ep_RAF_RAF_mod_RAFi_single_ddG);"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To implement the protection of the second dimer, we add additional energy patterns that change the Gibbs free energy of RAFi-RAF-RAF-RAFi quatramers."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "Parameter('ep_RAF_RAF_mod_RAFi_double_ddG', 1000.0)\n",
    "for RAF1, RAF2 in itt.combinations_with_replacement([BRAF, CRAF], r=2):\n",
    "    EnergyPattern(f'ep_{RAF1.name}_{RAF2.name}_mod_RAFi_double', \n",
    "                  RAFi(raf=2) % BRAF(raf=1, rafi=2) % BRAF(raf=1, rafi=3) % RAFi(raf=3), \n",
    "                  ep_RAF_RAF_mod_RAFi_double_ddG);"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will now load the remainder of the model from a file. Briefly, the full model describes MEK and ERK phosphorylation downstream as well as EGF stimulatable EGFR signaling upstream of RAF signaling. Moreover, it incorporates negative feedback from ERK on both MAPK and EGFR signaling. A comprehensive description of this part of the model is available in REF. Note however, that parts of the model were substantially simplified account for the reduced set of experimental data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "code_folding": [],
    "hide_input": false
   },
   "outputs": [],
   "source": [
    "from pysb import Observable, ANY\n",
    "def extend_model():\n",
    "    Monomer('RAS', ['raf', 'state'], {'state': ['gdp', 'gtp']})\n",
    "    Monomer('MEK', ['phospho'], {'phospho': ['p', 'u']})\n",
    "    Monomer('ERK', ['phospho'], {'phospho': ['p', 'u']})\n",
    "    Monomer('DUSP', [])\n",
    "\n",
    "    Parameter('ep_RAF_RAF_mod_RASstategtp_double_ddG', 1000.0)\n",
    "    Parameter('bind_RASstategtp_RAF_Ea', 10.0)\n",
    "    Parameter('bind_RASstategtp_RAF_dG', 0.01)\n",
    "    Parameter('bind_RASstategtp_RAF_phi', 1.0)\n",
    "    Parameter('catalyze_RAF_RAFrafiNone_MEK_p_kcat', 10.0)\n",
    "    Parameter('catalyze_RAFrafiNone_MEK_p_kcat', 10.0)\n",
    "    Parameter('catalyze_MEKmekiNone_phosphop_ERK_p_kcat', 10.0)\n",
    "    Parameter('synthesize_ERKphosphop_DUSP_ERK_gexpslope', 1000.0)\n",
    "    Parameter('synthesize_ERKphosphop_DUSP_kdeg', 10.0)\n",
    "    Parameter('synthesize_ERKphosphop_DUSP_ERK_kM', 0.0)\n",
    "    Parameter('inhibition_ERKphosphop_RAS_kM', 0.0)\n",
    "    Parameter('catalyze_PP2A_MEK_u_kcatr', 1.0)\n",
    "    Parameter('catalyze_DUSP_ERK_u_kcatr', 1.0)\n",
    "    Parameter('activation_RAS_kcat', 1.0)\n",
    "\n",
    "    Parameter('RAS_0', 0.0)\n",
    "    Parameter('MEK_0', 0.0)\n",
    "    Parameter('ERK_0', 0.0)\n",
    "    Parameter('EGF_0', 0.0)\n",
    "    Parameter('DUSP_eq', 10000.0)\n",
    "\n",
    "    Expression('Ea0_bind_RASstategtp_BRAF', \n",
    "               -bind_RASstategtp_RAF_phi*bind_RASstategtp_RAF_dG - bind_RASstategtp_RAF_Ea)\n",
    "    Expression('Ea0_bind_RASstategtp_CRAF', \n",
    "               -bind_RASstategtp_RAF_phi*bind_RASstategtp_RAF_dG - bind_RASstategtp_RAF_Ea)\n",
    "    \n",
    "    Expression('catalyze_PP2A_MEK_u_kcat', catalyze_PP2A_MEK_u_kcatr*catalyze_RAFrafiNone_MEK_p_kcat)\n",
    "    Expression('catalyze_DUSP_ERK_u_kcat', catalyze_DUSP_ERK_u_kcatr*catalyze_MEKmekiNone_phosphop_ERK_p_kcat)\n",
    "\n",
    "    Expression('initMEK', 1000000.0*MEK_0/(N_Avogadro*volume))\n",
    "    Expression('initERK', 1000000.0*ERK_0/(N_Avogadro*volume))\n",
    "    Expression('initRAS', 1000000.0*RAS_0/(N_Avogadro*volume))\n",
    "\n",
    "    Observable('modulation_ERKphosphop', ERK(phospho='p'))\n",
    "\n",
    "    Expression('synthesize_ERKphosphop_DUSP_ERK_kmod', modulation_ERKphosphop*synthesize_ERKphosphop_DUSP_ERK_gexpslope/(modulation_ERKphosphop + synthesize_ERKphosphop_DUSP_ERK_kM) + 1)\n",
    "    Expression('synthesize_ERKphosphop_DUSP_ksyn', 1.0*DUSP_eq*synthesize_ERKphosphop_DUSP_kdeg*synthesize_ERKphosphop_DUSP_ERK_kmod)\n",
    "    Expression('inhibition_ERKphosphop_RAS_kmod', activation_RAS_kcat/(1 + modulation_ERKphosphop/inhibition_ERKphosphop_RAS_kM)*EGF_0/100.0)\n",
    "\n",
    "\n",
    "    Rule('RASgtp_and_BRAF_bind_and_dissociate', \n",
    "         RAS(raf=None) + BRAF(RBD=None) | \n",
    "         RAS(raf=1) % BRAF(RBD=1), \n",
    "         bind_RASstategtp_RAF_phi, Ea0_bind_RASstategtp_BRAF, energy=True)\n",
    "    Rule('RASgtp_and_CRAF_bind_and_dissociate', \n",
    "         RAS(raf=None) + CRAF(RBD=None) | \n",
    "         RAS(raf=1) % CRAF(RBD=1), \n",
    "         bind_RASstategtp_RAF_phi, Ea0_bind_RASstategtp_CRAF, energy=True)\n",
    "    Rule('BRAF_BRAF_phosphorylates_MEK', \n",
    "         MEK(phospho='u') + BRAF(RBD=ANY, raf=2, rafi=None) % BRAF(RBD=ANY, raf=2) >> \n",
    "         MEK(phospho='p') + BRAF(RBD=ANY, raf=2, rafi=None) % BRAF(RBD=ANY, raf=2), \n",
    "         catalyze_RAF_RAFrafiNone_MEK_p_kcat)\n",
    "    Rule('CRAF_BRAF_phosphorylates_MEK', \n",
    "         MEK(phospho='u') + CRAF(RBD=ANY, raf=2, rafi=None) % BRAF(RBD=ANY, raf=2) >> \n",
    "         MEK(phospho='p') + CRAF(RBD=ANY, raf=2, rafi=None) % BRAF(RBD=ANY, raf=2), \n",
    "         catalyze_RAF_RAFrafiNone_MEK_p_kcat)\n",
    "    Rule('CRAF_CRAF_phosphorylates_MEK', \n",
    "         MEK(phospho='u') + CRAF(RBD=ANY, raf=2, rafi=None) % CRAF(RBD=ANY, raf=2) >> \n",
    "         MEK(phospho='p') + CRAF(RBD=ANY, raf=2, rafi=None) % CRAF(RBD=ANY, raf=2), \n",
    "         catalyze_RAF_RAFrafiNone_MEK_p_kcat)\n",
    "    Rule('BRAFV600E_phosphorylates_MEK_bound1', \n",
    "         MEK(phospho='u') + BRAF(AA600='E', raf=None, rafi=None) >> \n",
    "         MEK(phospho='p') + BRAF(AA600='E', raf=None, rafi=None), \n",
    "         catalyze_RAFrafiNone_MEK_p_kcat)\n",
    "    Rule('BRAFV600E_phosphorylates_MEK_bound2', \n",
    "         MEK(phospho='u') + BRAF(AA600='E', RBD=None, raf=ANY, rafi=None) >> \n",
    "         MEK(phospho='p') + BRAF(AA600='E', RBD=None, raf=ANY, rafi=None), \n",
    "         catalyze_RAFrafiNone_MEK_p_kcat)\n",
    "    Rule('BRAFV600E_phosphorylates_MEK_bound3', \n",
    "         MEK(phospho='u') + BRAF(AA600='E', RBD=ANY, raf=2, rafi=None) % BRAF(RBD=None, raf=2) >>\n",
    "         MEK(phospho='p') + BRAF(AA600='E', RBD=ANY, raf=2, rafi=None) % BRAF(RBD=None, raf=2), \n",
    "         catalyze_RAFrafiNone_MEK_p_kcat)\n",
    "    Rule('BRAFV600E_phosphorylates_MEK_bound4', \n",
    "         MEK(phospho='u') + BRAF(AA600='E', RBD=ANY, raf=2, rafi=None) % CRAF(RBD=None, raf=2) >> \n",
    "         MEK(phospho='p') + BRAF(AA600='E', RBD=ANY, raf=2, rafi=None) % CRAF(RBD=None, raf=2), \n",
    "         catalyze_RAFrafiNone_MEK_p_kcat)\n",
    "    Rule('MEK_is_dephosphorylated', \n",
    "         MEK(phospho='p') >> \n",
    "         MEK(phospho='u'), catalyze_PP2A_MEK_u_kcat)\n",
    "    Rule('pMEK_phosphorylates_ERK', \n",
    "         ERK(phospho='u') + MEK(phospho='p') >> \n",
    "         ERK(phospho='p') + MEK(phospho='p'), catalyze_MEKmekiNone_phosphop_ERK_p_kcat)\n",
    "    Rule('DUSP_dephosphorylates_ERK', \n",
    "         ERK(phospho='p') + DUSP() >> \n",
    "         ERK(phospho='u') + DUSP(), catalyze_DUSP_ERK_u_kcat)\n",
    "    Rule('synthesis_DUSP', \n",
    "         None >> DUSP(), \n",
    "         synthesize_ERKphosphop_DUSP_ksyn)\n",
    "    Rule('basal_degradation_DUSP', \n",
    "         DUSP() >> None, \n",
    "         synthesize_ERKphosphop_DUSP_kdeg, delete_molecules=True)\n",
    "    Rule('activation_RAS', \n",
    "         RAS(state='gdp') | RAS(state='gtp'), \n",
    "         inhibition_ERKphosphop_RAS_kmod, activation_RAS_kcat)\n",
    "\n",
    "    EnergyPattern('ep_BRAF_BRAF_mod_RAS_double', RAS(raf=2, state='gtp') % BRAF(RBD=2, raf=1) % BRAF(RBD=3, raf=1) % RAS(raf=3, state='gtp'), ep_RAF_RAF_mod_RASstategtp_double_ddG)\n",
    "    EnergyPattern('ep_BRAF_CRAF_mod_RAS_double', RAS(raf=2, state='gtp') % BRAF(RBD=2, raf=1) % CRAF(RBD=3, raf=1) % RAS(raf=3, state='gtp'), ep_RAF_RAF_mod_RASstategtp_double_ddG)\n",
    "    EnergyPattern('ep_CRAF_CRAF_mod_RAS_double', RAS(raf=2, state='gtp') % CRAF(RBD=2, raf=1) % CRAF(RBD=3, raf=1) % RAS(raf=3, state='gtp'), ep_RAF_RAF_mod_RASstategtp_double_ddG)\n",
    "    EnergyPattern('ep_bind_RASstategtp_BRAF', RAS(raf=1, state='gtp') % BRAF(RBD=1), bind_RASstategtp_RAF_dG)\n",
    "    EnergyPattern('ep_bind_RASstategtp_CRAF', RAS(raf=1, state='gtp') % CRAF(RBD=1), bind_RASstategtp_RAF_dG)\n",
    "\n",
    "    Initial(RAS(raf=None, state='gdp'), initRAS)\n",
    "    Initial(MEK(phospho='u'), initMEK)\n",
    "    Initial(ERK(phospho='u'), initERK)\n",
    "\n",
    "extend_model();"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Importing data in PEtab Format"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "With the model at hand, the next step for model calbitration is defining the training data. We will implement this in PEtab, which simplifies the definition of multiple experimental conditions. The petab specification of a calibration problem consists of the tables describing model observables, experimental measurements, experimental conditions and model parameters. Additionally, tables describing the visualization of simulations and data may also be included. In the following we will guide the reader through the definition"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Observables\n",
    "\n",
    "We will start the petab definition by specifying the model observables. Observables define the the model quantities that were measured experimentally, here pMEK and pERK. We add respective observables to the model using pysb Observables:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "Observable('pMEK_obs', MEK(phospho='p'))\n",
    "Observable('pERK_obs', ERK(phospho='p'));"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "These observables quantify all MEK and ERK molecules that are phosphorylated on the site `phospho`, which accounts for phosphorylation on TS on MEK and on ERK.\n",
    "\n",
    "The model quantifies pMEK and pERK in concentrations, but measurements are noise corrupted, measured in fluorescence intensity and also include background fluorescence. As the scaling between concentrations and intensity and the amount of background signal is known, we include scaling and offset parameters in the petab observable definition. To account for noise corruption, we specify a single noise parameter as noise Formula, which corresponds to Gaussian noise distribution with the respective parameter as standard deviation. The prefix `noiseParameter1` indicates that the value of the parameter will be provided in the respective column of the measurements table."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>observableFormula</th>\n",
       "      <th>noiseFormula</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>observableId</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>pMEK</th>\n",
       "      <td>pMEK_scale*pMEK_obs + pMEK_offset</td>\n",
       "      <td>noiseParameter1_pMEK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pERK</th>\n",
       "      <td>pERK_scale*pERK_obs + pERK_offset</td>\n",
       "      <td>noiseParameter1_pERK</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              observableFormula          noiseFormula\n",
       "observableId                                                         \n",
       "pMEK          pMEK_scale*pMEK_obs + pMEK_offset  noiseParameter1_pMEK\n",
       "pERK          pERK_scale*pERK_obs + pERK_offset  noiseParameter1_pERK"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import petab\n",
    "import pandas as pd\n",
    "observable_table = pd.DataFrame([\n",
    "    {\n",
    "        petab.OBSERVABLE_ID: obs_id,\n",
    "        petab.OBSERVABLE_FORMULA: f'{obs_id}_scale*{obs_id}_obs + {obs_id}_offset',\n",
    "        petab.NOISE_FORMULA: f'noiseParameter1_{obs_id}'\n",
    "    } for obs_id in ['pMEK', 'pERK']\n",
    "]).set_index(petab.OBSERVABLE_ID)\n",
    "observable_table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Measurements and Conditions\n",
    "As experimental measurement, we will load one of the datasets from REF, which contains the response of A375 cells cultured for 24h at different concentrations of multiple RAF and MEK inhibitors. The response is measured as normalized phospho-MEK (pMEK) and phospho-ERK (pERK) abundances in both EGF stimulated and unstimulated cells after 5 minutes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "UPGRADE AVAILABLE\n",
      "\n",
      "A more recent version of the Synapse Client (2.4.0) is available. Your version (2.3.1) can be upgraded by typing:\n",
      "    pip install --upgrade synapseclient\n",
      "\n",
      "Python Synapse Client version 2.4.0 release notes\n",
      "\n",
      "https://python-docs.synapse.org/build/html/news.html\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Cell_Line</th>\n",
       "      <th>Vemurafenib</th>\n",
       "      <th>Dabrafenib</th>\n",
       "      <th>...</th>\n",
       "      <th>Std.7</th>\n",
       "      <th>Std.8</th>\n",
       "      <th>Std.9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A375</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.009442</td>\n",
       "      <td>0.019582</td>\n",
       "      <td>0.007556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A375</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.111256</td>\n",
       "      <td>0.079383</td>\n",
       "      <td>0.141893</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>A375</td>\n",
       "      <td>0</td>\n",
       "      <td>0.1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.007028</td>\n",
       "      <td>0.018578</td>\n",
       "      <td>0.024706</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>A375</td>\n",
       "      <td>0</td>\n",
       "      <td>0.1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.014206</td>\n",
       "      <td>0.046046</td>\n",
       "      <td>0.117991</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>56 rows × 46 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Cell_Line  Vemurafenib  Dabrafenib  ...     Std.7     Std.8     Std.9\n",
       "0       A375           -1         0.0  ...  0.009442  0.019582  0.007556\n",
       "1       A375           -1         0.0  ...  0.111256  0.079383  0.141893\n",
       "..       ...          ...         ...  ...       ...       ...       ...\n",
       "54      A375            0         0.1  ...  0.007028  0.018578  0.024706\n",
       "55      A375            0         0.1  ...  0.014206  0.046046  0.117991\n",
       "\n",
       "[56 rows x 46 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import synapseclient as sc\n",
    "pd.set_option('display.max_rows', 5)\n",
    "pd.set_option('display.max_columns', 6)\n",
    "\n",
    "syn = sc.Synapse()\n",
    "syn.login(silent=True);\n",
    "data_df = pd.read_csv(syn.get('syn22804081').path)\n",
    "data_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The data contains the dose response to multiple RAF an MEK inhibitors as well as respective combinations. However, to minimize computation time requirements, we will only consider data involving the two RAFis vemurafenib and dabrafenib. The model itself only contains generic `RAFi` parameters and species, which we can now map to specific inhibitors using experimental conditions. This permits the simultaneous estimation of inhibitor specific kinetic rates in conjunction with all other model parameters. Such an multi-experiment setup can improve parameter identifiability. \n",
    "\n",
    "In the following, we will simulatenously generate the condition and measurement table. The condition table describes the necessary information about experiment, i.e., what drug was used, encoded in parameter mappings, and what concentrations were applied. The measurement table describes the measurement itself as well as additional information such as under which experimental conditions the data was measured.\n",
    "\n",
    "The data is encoded in a matrix format, while PEtab requires a long format. Thus, additional processing is necessary. Briefly, the code below extracts all data points and sets what model observable they belong to, the value of the measurement, the time of measurement, the noise parameter, information about the experimental condition as well as a dataset ID, which is only used for plotting.\n",
    "\n",
    "For the EGF stimulation condition, it is important to consider the experimental setup, as cells are subjected to 24h of pretreatment with drugs. We here assume that this is sufficient time for the system to reach a steady-state. To describe such an experimental setup in PEtab, it is necessary to reference the respective experimental condition in the condition table by specifying the respective `petab.CONDITION_ID` in the `petab.PREEQUILIBRATION_CONDITION_ID` column. The condition for the actual experiment can then be specified in the `petab.SIMULATION_CONDITION_ID`.\n",
    "\n",
    "To enable the visualization of data and fitting results, we here also specify the `petab.DATASET_ID`. Contrary to what the name may suggests, this does not influence the actual fitting, but serves as grouping identifier during plotting."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "RAFis = ['Vemurafenib', 'Dabrafenib', 'PLX8394', 'LY3009120', 'AZ_628']\n",
    "MEKis = ['Cobimetinib', 'Trametinib', 'Selumetinib', 'Binimetinib', 'PD0325901']\n",
    "\n",
    "\n",
    "conditions = dict()\n",
    "\n",
    "\n",
    "def format_petab(row):\n",
    "    suffixes = [''] + [f'.{idx}' for idx in range(1,10)]\n",
    "    datapoints = []\n",
    "    if row[MEKis].any():\n",
    "        return datapoints\n",
    "    # loop over columns of the data matrix\n",
    "    for suffix in suffixes:\n",
    "        # extract data\n",
    "        datapoint = {\n",
    "            petab.OBSERVABLE_ID: 'pMEK' if row.pMEK else 'pERK',\n",
    "            petab.MEASUREMENT: row[f'Mean{suffix}'],\n",
    "            petab.TIME: row['Time_EGF'],\n",
    "            petab.NOISE_PARAMETERS: row[f'Std{suffix}'],\n",
    "        }\n",
    "        # extract condition information\n",
    "        if (row[RAFis] != 0).sum() != 1 or ((row['Vemurafenib'] == 0) and (row['Dabrafenib'] == 0)):\n",
    "            continue\n",
    "\n",
    "        # find first nonzero\n",
    "        rafi = RAFis[(row[RAFis] != 0).argmax()]\n",
    "\n",
    "        # extract drug concentration\n",
    "        drug_conc = row[f'Concentration (uM){suffix}'] if row[rafi] == -1 else row[rafi]\n",
    "        # condition id must be sanitzed, must match '^[a-zA-Z]+[\\w_]*$'\n",
    "        drug_str = f'{rafi}_{drug_conc}'.replace('.', '_').replace('-', '_')\n",
    "\n",
    "        condition_str = drug_str + (\n",
    "            f'__EGF_{row[\"EGF\"]}' if row['EGF'] > 0 else ''\n",
    "        )\n",
    "        condition = {\n",
    "            petab.CONDITION_ID: condition_str,'RAFi_0': drug_conc,\n",
    "            'EGF_0': row[\"EGF\"], 'bind_RAFi_RAF_Ea': f'bind_{rafi}_RAF_Ea',\n",
    "            'bind_RAFi_RAF_dG': f'bind_{rafi}_RAF_dG',\n",
    "            'ep_RAF_RAF_mod_RAFi_single_ddG': f'ep_RAF_RAF_mod_{rafi}_single_ddG',\n",
    "            'ep_RAF_RAF_mod_RAFi_double_ddG': f'ep_RAF_RAF_mod_{rafi}_double_ddG'\n",
    "        }\n",
    "        # set baseline for datapoint\n",
    "        datapoint[petab.PREEQUILIBRATION_CONDITION_ID] = drug_str\n",
    "        # set id for condition and datapoint\n",
    "        datapoint[petab.SIMULATION_CONDITION_ID] = condition_str\n",
    "        datapoint[petab.DATASET_ID] =  ('EGF__' if row['EGF'] > 0 else 'ctrl__') + rafi\n",
    "        condition[petab.CONDITION_ID] = condition_str\n",
    "        \n",
    "        datapoints.append(datapoint)\n",
    "        if condition_str not in conditions:\n",
    "            conditions[condition_str] = condition\n",
    "            \n",
    "    return datapoints\n",
    "\n",
    "measurement_table = pd.DataFrame([\n",
    "    d\n",
    "    for ir, row in data_df.iterrows()\n",
    "    for d in format_petab(row)\n",
    "])\n",
    "condition_table = pd.DataFrame(conditions.values()).set_index(petab.CONDITION_ID);"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the above code we directly used experimentally measured standard deviations, which can be noisy as they are often computed from a small number of biological or technical replicates. As the magnitude of standard deviations determines the importance of datapoints during calibration, outliers can lead to overemphasis of individual datapoints resulting in poor fits to the overall data. To avoid such issues, we use the same averaged standard deviations for all datapoints of each observable."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "for group, frame in measurement_table.groupby(petab.OBSERVABLE_ID):\n",
    "    measurement_table.loc[measurement_table[petab.OBSERVABLE_ID]==group, petab.NOISE_PARAMETERS] = \\\n",
    "        np.sqrt(frame[petab.NOISE_PARAMETERS].apply(np.square).mean())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>observableId</th>\n",
       "      <th>measurement</th>\n",
       "      <th>time</th>\n",
       "      <th>...</th>\n",
       "      <th>preequilibrationConditionId</th>\n",
       "      <th>simulationConditionId</th>\n",
       "      <th>datasetId</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>pERK</td>\n",
       "      <td>0.969789</td>\n",
       "      <td>0.0833</td>\n",
       "      <td>...</td>\n",
       "      <td>Vemurafenib_0_000210819</td>\n",
       "      <td>Vemurafenib_0_000210819</td>\n",
       "      <td>ctrl__Vemurafenib</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>pERK</td>\n",
       "      <td>0.944607</td>\n",
       "      <td>0.0833</td>\n",
       "      <td>...</td>\n",
       "      <td>Vemurafenib_0_000666667</td>\n",
       "      <td>Vemurafenib_0_000666667</td>\n",
       "      <td>ctrl__Vemurafenib</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>pMEK</td>\n",
       "      <td>1.172071</td>\n",
       "      <td>0.0833</td>\n",
       "      <td>...</td>\n",
       "      <td>Dabrafenib_0_210818511</td>\n",
       "      <td>Dabrafenib_0_210818511__EGF_100</td>\n",
       "      <td>EGF__Dabrafenib</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>pMEK</td>\n",
       "      <td>1.091548</td>\n",
       "      <td>0.0833</td>\n",
       "      <td>...</td>\n",
       "      <td>Dabrafenib_0_666666667</td>\n",
       "      <td>Dabrafenib_0_666666667__EGF_100</td>\n",
       "      <td>EGF__Dabrafenib</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>80 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   observableId  measurement    time  ...  preequilibrationConditionId  \\\n",
       "0          pERK     0.969789  0.0833  ...      Vemurafenib_0_000210819   \n",
       "1          pERK     0.944607  0.0833  ...      Vemurafenib_0_000666667   \n",
       "..          ...          ...     ...  ...                          ...   \n",
       "78         pMEK     1.172071  0.0833  ...       Dabrafenib_0_210818511   \n",
       "79         pMEK     1.091548  0.0833  ...       Dabrafenib_0_666666667   \n",
       "\n",
       "              simulationConditionId          datasetId  \n",
       "0           Vemurafenib_0_000210819  ctrl__Vemurafenib  \n",
       "1           Vemurafenib_0_000666667  ctrl__Vemurafenib  \n",
       "..                              ...                ...  \n",
       "78  Dabrafenib_0_210818511__EGF_100    EGF__Dabrafenib  \n",
       "79  Dabrafenib_0_666666667__EGF_100    EGF__Dabrafenib  \n",
       "\n",
       "[80 rows x 7 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "measurement_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RAFi_0</th>\n",
       "      <th>EGF_0</th>\n",
       "      <th>bind_RAFi_RAF_Ea</th>\n",
       "      <th>bind_RAFi_RAF_dG</th>\n",
       "      <th>ep_RAF_RAF_mod_RAFi_single_ddG</th>\n",
       "      <th>ep_RAF_RAF_mod_RAFi_double_ddG</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>conditionId</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Vemurafenib_0_000210819</th>\n",
       "      <td>0.000211</td>\n",
       "      <td>0</td>\n",
       "      <td>bind_Vemurafenib_RAF_Ea</td>\n",
       "      <td>bind_Vemurafenib_RAF_dG</td>\n",
       "      <td>ep_RAF_RAF_mod_Vemurafenib_single_ddG</td>\n",
       "      <td>ep_RAF_RAF_mod_Vemurafenib_double_ddG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Vemurafenib_0_000666667</th>\n",
       "      <td>0.000667</td>\n",
       "      <td>0</td>\n",
       "      <td>bind_Vemurafenib_RAF_Ea</td>\n",
       "      <td>bind_Vemurafenib_RAF_dG</td>\n",
       "      <td>ep_RAF_RAF_mod_Vemurafenib_single_ddG</td>\n",
       "      <td>ep_RAF_RAF_mod_Vemurafenib_double_ddG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Dabrafenib_0_210818511__EGF_100</th>\n",
       "      <td>0.210819</td>\n",
       "      <td>100</td>\n",
       "      <td>bind_Dabrafenib_RAF_Ea</td>\n",
       "      <td>bind_Dabrafenib_RAF_dG</td>\n",
       "      <td>ep_RAF_RAF_mod_Dabrafenib_single_ddG</td>\n",
       "      <td>ep_RAF_RAF_mod_Dabrafenib_double_ddG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Dabrafenib_0_666666667__EGF_100</th>\n",
       "      <td>0.666667</td>\n",
       "      <td>100</td>\n",
       "      <td>bind_Dabrafenib_RAF_Ea</td>\n",
       "      <td>bind_Dabrafenib_RAF_dG</td>\n",
       "      <td>ep_RAF_RAF_mod_Dabrafenib_single_ddG</td>\n",
       "      <td>ep_RAF_RAF_mod_Dabrafenib_double_ddG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>40 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   RAFi_0  EGF_0         bind_RAFi_RAF_Ea  \\\n",
       "conditionId                                                                 \n",
       "Vemurafenib_0_000210819          0.000211      0  bind_Vemurafenib_RAF_Ea   \n",
       "Vemurafenib_0_000666667          0.000667      0  bind_Vemurafenib_RAF_Ea   \n",
       "...                                   ...    ...                      ...   \n",
       "Dabrafenib_0_210818511__EGF_100  0.210819    100   bind_Dabrafenib_RAF_Ea   \n",
       "Dabrafenib_0_666666667__EGF_100  0.666667    100   bind_Dabrafenib_RAF_Ea   \n",
       "\n",
       "                                        bind_RAFi_RAF_dG  \\\n",
       "conditionId                                                \n",
       "Vemurafenib_0_000210819          bind_Vemurafenib_RAF_dG   \n",
       "Vemurafenib_0_000666667          bind_Vemurafenib_RAF_dG   \n",
       "...                                                  ...   \n",
       "Dabrafenib_0_210818511__EGF_100   bind_Dabrafenib_RAF_dG   \n",
       "Dabrafenib_0_666666667__EGF_100   bind_Dabrafenib_RAF_dG   \n",
       "\n",
       "                                        ep_RAF_RAF_mod_RAFi_single_ddG  \\\n",
       "conditionId                                                              \n",
       "Vemurafenib_0_000210819          ep_RAF_RAF_mod_Vemurafenib_single_ddG   \n",
       "Vemurafenib_0_000666667          ep_RAF_RAF_mod_Vemurafenib_single_ddG   \n",
       "...                                                                ...   \n",
       "Dabrafenib_0_210818511__EGF_100   ep_RAF_RAF_mod_Dabrafenib_single_ddG   \n",
       "Dabrafenib_0_666666667__EGF_100   ep_RAF_RAF_mod_Dabrafenib_single_ddG   \n",
       "\n",
       "                                        ep_RAF_RAF_mod_RAFi_double_ddG  \n",
       "conditionId                                                             \n",
       "Vemurafenib_0_000210819          ep_RAF_RAF_mod_Vemurafenib_double_ddG  \n",
       "Vemurafenib_0_000666667          ep_RAF_RAF_mod_Vemurafenib_double_ddG  \n",
       "...                                                                ...  \n",
       "Dabrafenib_0_210818511__EGF_100   ep_RAF_RAF_mod_Dabrafenib_double_ddG  \n",
       "Dabrafenib_0_666666667__EGF_100   ep_RAF_RAF_mod_Dabrafenib_double_ddG  \n",
       "\n",
       "[40 rows x 6 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "condition_table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Parameters\n",
    "The last table necessary for model training is the parameter table. This table describes all free parameters of the model, whether they have to be estimated as well as, parameter scale, nominal values, boundary values and information about priors.\n",
    "\n",
    "Both parameter boundaries and priors can be used to restrict the search space to biologically plausible parameter values, or prevent numerical issues in regions of parameter space where numerical integration of model equations is challenging. Parameter boundaries set hard threshold that cannot be exceeded, while parameter boundaries are soft constraints that also regularize the calibration. While priors are necessary for Bayesian analysis such as parameter sampling, they are not generally required for optimization based approaches.\n",
    "\n",
    "For most models, free parameters will primarily occur in the model specification. But, all of previously described tables may introduce new parameters, e.g., in noise or observable formulas or condition. Moreover, the tables may fix the values of both model parameters and newly introduced parameters to numerical values, rendering them non-estimatable. As the parameter table must only describe free parameters, some processing is usually necessary to identify the set of parameters that require specification in the parameter table.\n",
    "\n",
    "In the following, we will construct a list of `condition_pars`, which are either mapped to  condition specific parameters, such as the generic RAFi energy parameters, or set to fixed values, such as the `RAFi_0` initial concentration. To construct the list of free parameters, we identify all parameters in the model that are not contained in these `condition_pars`, and add all newly introduced condition specific parameters in the condition table as well as the four offset and scaling parameters that were introduced in the observables table."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "condition_pars = [\n",
    "    par.name\n",
    "    for par in model.parameters\n",
    "    if par.name in condition_table.columns\n",
    "]\n",
    "\n",
    "free_parameters = [\n",
    "    par.name for par in model.parameters\n",
    "    if par.name not in condition_pars\n",
    "] + [\n",
    "    name\n",
    "    for par in condition_pars\n",
    "    for name in np.unique(condition_table[par])\n",
    "    if isinstance(name, str)\n",
    "] + [\n",
    "    'pMEK_offset','pMEK_scale', 'pERK_offset', 'pERK_scale'\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "All of the parameters in the model follow a consistent naming scheme, which makes it easier to programmatically specify parameter boundaries, scales and whether parameters are to be estimated. Multiple studies have demonstrated that parameter estimation is more efficient when estimating parmaters on a logarithmic scale. The thermodynamic formulation we employed for model construction relies on an exponential dependency between energies and kinetetic rates, which complements these practical insights with a theoretical foundation why model parameters live on a logarithmic scale. However, this also suggests that energetic parameters, in contrast to kinetic rates, should be estimated on a linear scale.\n",
    "\n",
    "The upper and lower boundaries should ideally be set to biologically plausible values. However, most models employ simplified descriptions of the underlying biochemical processes, which means that model parameters may no longer have a one-to-one correspondance to true biochemical constants and plausible parameter ranges may be difficult to derive. Moreover, plausible parameter ranges may not be known for all parameters. Therefore, in many applications parameter boundaries are initially based on educated guesses and then refined based on estimation results such that optima are contained in the specified ranges and integration failures are not too frequent. Therefore, the boundaries presented here are based on the values used in the original publication and where then refined to work with the simplified model.\n",
    "\n",
    "Similar to parameter boundaries, nominal values can be difficult to derive. However, the numerical values are only important for model calibration when respective parameters are not estimated. In this example, we do not estimate the initial concentrations and $\\phi$ parameters and only set the nominal parameter values of those parameters to non-trivial values. The initial concentrations are set to approximate molecular counts as measured in the original publication and $\\phi$ parameters are all set to 1, meaning that free energy differences only affect the reverse rate.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>parameterScale</th>\n",
       "      <th>lowerBound</th>\n",
       "      <th>upperBound</th>\n",
       "      <th>nominalValue</th>\n",
       "      <th>estimate</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>parameterId</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>BRAF_0</th>\n",
       "      <td>log10</td>\n",
       "      <td>1000.00</td>\n",
       "      <td>100000.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CRAF_0</th>\n",
       "      <td>log10</td>\n",
       "      <td>1000.00</td>\n",
       "      <td>100000.0</td>\n",
       "      <td>10000.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pERK_offset</th>\n",
       "      <td>log10</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pERK_scale</th>\n",
       "      <td>log10</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1000000.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>36 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            parameterScale  lowerBound  upperBound  nominalValue  estimate\n",
       "parameterId                                                               \n",
       "BRAF_0               log10     1000.00    100000.0        1000.0     False\n",
       "CRAF_0               log10     1000.00    100000.0       10000.0     False\n",
       "...                    ...         ...         ...           ...       ...\n",
       "pERK_offset          log10        0.05         0.2           1.0      True\n",
       "pERK_scale           log10        1.00   1000000.0           1.0      True\n",
       "\n",
       "[36 rows x 5 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lbs = {\n",
    "    'Ea': -10, 'dG': -10, 'ddG': -10, 'phi': 0, 'offset': 5e-2,\n",
    "    'scale': 1e0, 'eq': 1e1, '0': 1e3, 'kcat': 1e-2, 'gexpslope': 1e0,\n",
    "    'kdeg': 1e-3, 'kM': 1e-3, 'kcatr': 1e-8, 'koff': 1e-5,\n",
    "}\n",
    "ubs = {\n",
    "    'Ea': 10, 'dG': 10, 'ddG': 10, 'phi': 1, 'offset': 2e-1,\n",
    "    'scale': 1e6, 'eq': 1e5, '0': 1e5, 'kcat': 1e5, 'gexpslope': 1e6,\n",
    "    'kdeg': 1e2, 'kM': 1e1, 'kcatr': 1e0, 'koff': 1e0,\n",
    "}\n",
    "\n",
    "initials = {\n",
    "    'BRAF_0': 1e3, 'CRAF_0': 1e4, 'MEK_0': 1e5, 'ERK_0': 1e5, 'RAS_0': 5e4,\n",
    "}\n",
    "\n",
    "\n",
    "parameter_table = pd.DataFrame([\n",
    "    {\n",
    "        petab.PARAMETER_ID: par,\n",
    "        petab.PARAMETER_SCALE: petab.LIN \n",
    "        if par.endswith(('_Ea', '_dG', '_ddG', '_phi')) else petab.LOG10,\n",
    "        petab.LOWER_BOUND: lbs[par.split('_')[-1]],\n",
    "        petab.UPPER_BOUND: ubs[par.split('_')[-1]],\n",
    "        petab.NOMINAL_VALUE: initials[par] if par.endswith('_0') else 1.0,\n",
    "        petab.ESTIMATE: False if par.endswith(('_phi', '_0')) else True,\n",
    "        \n",
    "    } for par in free_parameters\n",
    "]).set_index(petab.PARAMETER_ID)\n",
    "\n",
    "parameter_table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## SBML export\n",
    "PEtab expects a model in SBML format, but we constructed the model in the pysb format. Therefore, we need to export the model in SBML format."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "import libsbml\n",
    "from pysb.export import export\n",
    "sbml_reader = libsbml.SBMLReader()\n",
    "sbml_doc = sbml_reader.readSBMLFromString(export(model, 'sbml'))\n",
    "sbml_model = sbml_doc.getModel()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the SBML model, pysb observables are exported as SBML parameters with assignment rules, which programmatically generated identifiers that do not agree with the observable names. As we specified observable formulas based on the respective pysb names, we have to change the identifiers of these parameter to match the pysb names such that the respective references are recognized during petab import."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# rename pysb exported observables\n",
    "sbml_model.getParameter('__obs1').setId('pMEK_obs')\n",
    "sbml_model.getAssignmentRuleByVariable('__obs1').setVariable('pMEK_obs')\n",
    "sbml_model.getParameter('__obs2').setId('pERK_obs')\n",
    "sbml_model.getAssignmentRuleByVariable('__obs2').setVariable('pERK_obs');"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As last step, we create a petab problem using all of the previously constructed table as well as the exported SBML model. As problem construction was a lengthy error-prone process, we performt a static analysis of the petab problem using the petab library provided linter."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s6' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s7' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s8' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s9' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s10' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s11' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s12' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s13' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s14' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s15' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s16' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s17' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s18' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s19' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s20' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s21' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s22' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s23' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s24' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s25' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s26' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s27' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s28' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s29' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s30' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s31' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s32' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s33' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s34' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s35' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s36' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s37' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s38' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s39' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s40' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s41' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s42' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s43' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s44' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s45' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s46' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s47' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s48' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s49' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s50' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s51' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s52' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s53' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s54' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s55' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s56' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s57' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s58' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s59' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s60' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s61' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s62' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s63' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s64' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s65' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s66' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s67' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s68' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s69' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s70' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s71' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s72' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s73' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s74' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s75' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s76' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s77' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s78' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s79' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s80' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s81' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s82' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s83' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s84' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s85' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s86' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s87' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s88' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s89' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s90' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s91' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s92' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s93' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s94' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s95' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s96' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s97' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "WARNING: Generated invalid SBML model. Check messages above.\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s6' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s7' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s8' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s9' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s10' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s11' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s12' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s13' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s14' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s15' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s16' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s17' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s18' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s19' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s20' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s21' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s22' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s23' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s24' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s25' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s26' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s27' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s28' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s29' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s30' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s31' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s32' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s33' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s34' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s35' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s36' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s37' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s38' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s39' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s40' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s41' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s42' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s43' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s44' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s45' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s46' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s47' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s48' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s49' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s50' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s51' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s52' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s53' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s54' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s55' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s56' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s57' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s58' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s59' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s60' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s61' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s62' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s63' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s64' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s65' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s66' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s67' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s68' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s69' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s70' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s71' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s72' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s73' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s74' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s75' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s76' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s77' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s78' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s79' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s80' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s81' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s82' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s83' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s84' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s85' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s86' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s87' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s88' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s89' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s90' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s91' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s92' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s93' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s94' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s95' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s96' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "libSBML Warning (Modeling practice): As a principle of best modeling practice, the <species> should set an initial value (amount or concentration) rather than be left undefined. Doing so improves the portability of models between different simulation and analysis systems, and helps make it easier to detect potential errors in models.\n",
      " The <species> with the id '__s97' does not have an 'initialConcentration' or 'initialAmount' attribute, nor is its initial value set by an <initialAssignment> or <assignmentRule>.\n",
      "\n",
      "Not OK\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import petab\n",
    "petab_problem = petab.Problem(\n",
    "    sbml_model=sbml_model, sbml_reader=sbml_reader, sbml_document=sbml_doc,\n",
    "    measurement_df=measurement_table, condition_df=condition_table,\n",
    "    observable_df=observable_table, parameter_df=parameter_table\n",
    ")\n",
    "petab.lint.lint_problem(petab_problem)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Calibrating the model in pyPESTO\n",
    "Model calibration based on PEtab format is supported by a growing number of tools including COPASI and data2dynamics. Here we used the python based calibration tool pyPESTO for calibration. pyPESTO uses AMICI for model import and simulation, constructs an objective function and provides an interface to various different optimizers. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Compiling amici model to folder /Users/ffroehlich/Documents/HMS/energy_modeling_chapter/amici_models/thermo_raf.\n",
      "2021-08-05 11:54:14.744 - amici.petab_import - INFO - Importing model ...\n",
      "2021-08-05 11:54:14.746 - amici.petab_import - INFO - Model name is 'thermo_raf'.\n",
      "Writing model code to '/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/amici_models/thermo_raf'.\n",
      "2021-08-05 11:54:14.762 - amici.petab_import - INFO - Species: 98\n",
      "2021-08-05 11:54:14.763 - amici.petab_import - INFO - Global parameters: 117\n",
      "2021-08-05 11:54:14.763 - amici.petab_import - INFO - Reactions: 360\n",
      "2021-08-05 11:54:14.835 - amici.petab_import - INFO - Observables: 2\n",
      "2021-08-05 11:54:14.836 - amici.petab_import - INFO - Sigmas: 2\n",
      "2021-08-05 11:54:14.840 - amici.petab_import - DEBUG - Adding output parameters to model: OrderedDict([('pMEK_offset', None), ('pMEK_scale', None), ('pERK_offset', None), ('pERK_scale', None), ('noiseParameter1_pMEK', None), ('noiseParameter1_pERK', None)])\n",
      "2021-08-05 11:54:14.841 - amici.petab_import - DEBUG - Adding initial assignments for []\n",
      "2021-08-05 11:54:14.842 - amici.petab_import - DEBUG - Condition table: (40, 6)\n",
      "2021-08-05 11:54:14.843 - amici.petab_import - DEBUG - Fixed parameters are ['RAFi_0', 'EGF_0']\n",
      "2021-08-05 11:54:14.844 - amici.petab_import - INFO - Overall fixed parameters: 2\n",
      "2021-08-05 11:54:14.845 - amici.petab_import - INFO - Variable parameters: 121\n",
      "2021-08-05 11:54:15.429 - amici.sbml_import - INFO - Finished gathering local SBML symbols       ++ (5.69E-01s)\n",
      "2021-08-05 11:54:15.588 - amici.sbml_import - INFO - Finished processing SBML parameters         ++ (1.54E-01s)\n",
      "2021-08-05 11:54:15.594 - amici.sbml_import - INFO - Finished processing SBML compartments       ++ (2.79E-04s)\n",
      "2021-08-05 11:54:15.642 - amici.sbml_import - INFO - Finished processing SBML species initials  +++ (3.04E-02s)\n",
      "2021-08-05 11:54:15.648 - amici.sbml_import - INFO - Finished processing SBML rate rules        +++ (3.35E-04s)\n",
      "2021-08-05 11:54:15.649 - amici.sbml_import - INFO - Finished processing SBML species            ++ (4.63E-02s)\n",
      "2021-08-05 11:54:16.002 - amici.sbml_import - INFO - Finished processing SBML reactions          ++ (3.47E-01s)\n",
      "2021-08-05 11:54:16.097 - amici.sbml_import - INFO - Finished processing SBML rules              ++ (8.98E-02s)\n",
      "2021-08-05 11:54:16.101 - amici.sbml_import - INFO - Finished processing SBML initial assignments++ (1.94E-04s)\n",
      "2021-08-05 11:54:16.128 - amici.sbml_import - INFO - Finished processing SBML species references ++ (2.23E-02s)\n",
      "2021-08-05 11:54:16.133 - amici.sbml_import - INFO - Finished processing SBML events             ++ (8.71E-05s)\n",
      "2021-08-05 11:54:16.134 - amici.sbml_import - INFO - Finished importing SBML                      + (1.28E+00s)\n",
      "2021-08-05 11:54:16.161 - amici.sbml_import - INFO - Finished processing SBML observables         + (2.16E-02s)\n",
      "2021-08-05 11:54:17.016 - amici.ode_export - INFO - Finished running smart_multiply              ++ (1.84E-01s)\n",
      "2021-08-05 11:54:21.837 - amici.ode_export - INFO - Finished importing SbmlImporter               + (5.02E+00s)\n",
      "2021-08-05 11:54:21.868 - amici.ode_export - INFO - Finished simplifying Jy                    ++++ (1.56E-02s)\n",
      "2021-08-05 11:54:21.869 - amici.ode_export - INFO - Finished computing Jy                       +++ (2.02E-02s)\n",
      "2021-08-05 11:54:21.880 - amici.ode_export - INFO - Finished simplifying y                     ++++ (1.52E-03s)\n",
      "2021-08-05 11:54:21.881 - amici.ode_export - INFO - Finished computing y                        +++ (5.42E-03s)\n",
      "2021-08-05 11:54:21.893 - amici.ode_export - INFO - Finished simplifying sigmay                ++++ (1.00E-04s)\n",
      "2021-08-05 11:54:21.894 - amici.ode_export - INFO - Finished computing sigmay                   +++ (6.21E-03s)\n",
      "2021-08-05 11:54:21.937 - amici.ode_export - INFO - Finished writing Jy.cpp                      ++ (9.06E-02s)\n",
      "2021-08-05 11:54:21.993 - amici.ode_export - INFO - Finished running smart_jacobian            ++++ (3.18E-02s)\n",
      "2021-08-05 11:54:22.018 - amici.ode_export - INFO - Finished simplifying dJydsigma             ++++ (1.72E-02s)\n",
      "2021-08-05 11:54:22.019 - amici.ode_export - INFO - Finished computing dJydsigma                +++ (6.44E-02s)\n",
      "2021-08-05 11:54:22.027 - amici.ode_export - INFO - Finished writing dJydsigma.cpp               ++ (7.89E-02s)\n",
      "2021-08-05 11:54:22.058 - amici.ode_export - INFO - Finished running smart_jacobian            ++++ (1.42E-02s)\n",
      "2021-08-05 11:54:22.079 - amici.ode_export - INFO - Finished simplifying dJydy                 ++++ (1.26E-02s)\n",
      "2021-08-05 11:54:22.080 - amici.ode_export - INFO - Finished computing dJydy                    +++ (4.00E-02s)\n",
      "2021-08-05 11:54:22.091 - amici.ode_export - INFO - Finished writing dJydy.cpp                   ++ (5.61E-02s)\n",
      "2021-08-05 11:54:22.107 - amici.ode_export - INFO - Finished simplifying root                  ++++ (1.15E-04s)\n",
      "2021-08-05 11:54:22.108 - amici.ode_export - INFO - Finished computing root                     +++ (6.72E-03s)\n",
      "2021-08-05 11:54:22.109 - amici.ode_export - INFO - Finished writing root.cpp                    ++ (1.12E-02s)\n",
      "2021-08-05 11:54:22.628 - amici.ode_export - INFO - Finished simplifying w                   ++++++ (4.91E-01s)\n",
      "2021-08-05 11:54:22.628 - amici.ode_export - INFO - Finished computing w                      +++++ (4.98E-01s)\n",
      "2021-08-05 11:54:24.034 - amici.ode_export - INFO - Finished running smart_jacobian           +++++ (1.40E+00s)\n",
      "2021-08-05 11:54:25.081 - amici.ode_export - INFO - Finished simplifying dwdp                 +++++ (1.04E+00s)\n",
      "2021-08-05 11:54:25.083 - amici.ode_export - INFO - Finished computing dwdp                    ++++ (2.96E+00s)\n",
      "2021-08-05 11:54:25.153 - amici.ode_export - INFO - Finished running smart_jacobian           +++++ (4.18E-02s)\n",
      "2021-08-05 11:54:25.161 - amici.ode_export - INFO - Finished simplifying dwdtcl               +++++ (1.67E-03s)\n",
      "2021-08-05 11:54:25.162 - amici.ode_export - INFO - Finished computing dwdtcl                  ++++ (5.64E-02s)\n",
      "2021-08-05 11:54:26.005 - amici.ode_export - INFO - Finished simplifying dwdp                  ++++ (8.34E-01s)\n",
      "2021-08-05 11:54:26.007 - amici.ode_export - INFO - Finished computing dwdp                     +++ (3.89E+00s)\n",
      "2021-08-05 11:54:26.711 - amici.ode_export - INFO - Finished writing dwdp.cpp                    ++ (4.60E+00s)\n",
      "2021-08-05 11:54:29.802 - amici.ode_export - INFO - Finished running smart_jacobian            ++++ (3.08E+00s)\n",
      "2021-08-05 11:54:30.348 - amici.ode_export - INFO - Finished simplifying dwdx                  ++++ (5.40E-01s)\n",
      "2021-08-05 11:54:30.349 - amici.ode_export - INFO - Finished computing dwdx                     +++ (3.63E+00s)\n",
      "2021-08-05 11:54:31.418 - amici.ode_export - INFO - Finished writing dwdx.cpp                    ++ (4.70E+00s)\n",
      "2021-08-05 11:54:40.828 - amici.ode_export - INFO - Finished running smart_jacobian            ++++ (9.39E+00s)\n",
      "2021-08-05 11:54:42.741 - amici.ode_export - INFO - Finished simplifying dwdw                  ++++ (7.85E-01s)\n",
      "2021-08-05 11:54:42.743 - amici.ode_export - INFO - Finished computing dwdw                     +++ (1.13E+01s)\n",
      "2021-08-05 11:54:43.831 - amici.ode_export - INFO - Finished writing dwdw.cpp                    ++ (1.24E+01s)\n",
      "2021-08-05 11:54:44.320 - amici.ode_export - INFO - Finished simplifying xdot                  ++++ (4.78E-01s)\n",
      "2021-08-05 11:54:44.320 - amici.ode_export - INFO - Finished computing xdot                     +++ (4.82E-01s)\n",
      "2021-08-05 11:54:45.030 - amici.ode_export - INFO - Finished writing dxdotdw.cpp                 ++ (1.19E+00s)\n",
      "2021-08-05 11:54:45.045 - amici.ode_export - INFO - Finished running smart_jacobian            ++++ (5.72E-03s)\n",
      "2021-08-05 11:54:45.080 - amici.ode_export - INFO - Finished simplifying dxdotdx_explicit      ++++ (2.96E-02s)\n",
      "2021-08-05 11:54:45.080 - amici.ode_export - INFO - Finished computing dxdotdx_explicit         +++ (4.31E-02s)\n",
      "2021-08-05 11:54:45.119 - amici.ode_export - INFO - Finished writing dxdotdx_explicit.cpp        ++ (8.41E-02s)\n",
      "2021-08-05 11:54:45.137 - amici.ode_export - INFO - Finished running smart_jacobian            ++++ (1.77E-03s)\n",
      "2021-08-05 11:54:45.152 - amici.ode_export - INFO - Finished simplifying dxdotdp_explicit      ++++ (1.03E-02s)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2021-08-05 11:54:45.153 - amici.ode_export - INFO - Finished computing dxdotdp_explicit         +++ (2.31E-02s)\n",
      "2021-08-05 11:54:45.167 - amici.ode_export - INFO - Finished writing dxdotdp_explicit.cpp        ++ (4.06E-02s)\n",
      "2021-08-05 11:54:45.187 - amici.ode_export - INFO - Finished running smart_jacobian           +++++ (4.59E-04s)\n",
      "2021-08-05 11:54:45.194 - amici.ode_export - INFO - Finished simplifying dydx                 +++++ (7.41E-04s)\n",
      "2021-08-05 11:54:45.195 - amici.ode_export - INFO - Finished computing dydx                    ++++ (1.16E-02s)\n",
      "2021-08-05 11:54:45.246 - amici.ode_export - INFO - Finished running smart_jacobian           +++++ (3.94E-02s)\n",
      "2021-08-05 11:54:45.651 - amici.ode_export - INFO - Finished running smart_multiply           +++++ (4.00E-01s)\n",
      "2021-08-05 11:54:45.662 - amici.ode_export - INFO - Finished simplifying dydw                 +++++ (3.21E-03s)\n",
      "2021-08-05 11:54:45.663 - amici.ode_export - INFO - Finished computing dydw                    ++++ (4.62E-01s)\n",
      "2021-08-05 11:54:45.766 - amici.ode_export - INFO - Finished running smart_multiply            ++++ (9.70E-02s)\n",
      "2021-08-05 11:54:45.774 - amici.ode_export - INFO - Finished simplifying dydx                  ++++ (7.42E-04s)\n",
      "2021-08-05 11:54:45.775 - amici.ode_export - INFO - Finished computing dydx                     +++ (5.96E-01s)\n",
      "2021-08-05 11:54:45.778 - amici.ode_export - INFO - Finished writing dydx.cpp                    ++ (6.04E-01s)\n",
      "2021-08-05 11:54:45.797 - amici.ode_export - INFO - Finished running smart_jacobian           +++++ (5.10E-03s)\n",
      "2021-08-05 11:54:45.804 - amici.ode_export - INFO - Finished simplifying dydp                 +++++ (4.76E-04s)\n",
      "2021-08-05 11:54:45.805 - amici.ode_export - INFO - Finished computing dydp                    ++++ (1.65E-02s)\n",
      "2021-08-05 11:54:45.860 - amici.ode_export - INFO - Finished running smart_multiply            ++++ (4.57E-02s)\n",
      "2021-08-05 11:54:45.864 - amici.ode_export - INFO - Finished simplifying dydp                  ++++ (2.42E-04s)\n",
      "2021-08-05 11:54:45.865 - amici.ode_export - INFO - Finished computing dydp                     +++ (7.97E-02s)\n",
      "2021-08-05 11:54:45.868 - amici.ode_export - INFO - Finished writing dydp.cpp                    ++ (8.50E-02s)\n",
      "2021-08-05 11:54:45.882 - amici.ode_export - INFO - Finished running smart_jacobian            ++++ (2.39E-03s)\n",
      "2021-08-05 11:54:45.888 - amici.ode_export - INFO - Finished simplifying dsigmaydp             ++++ (4.62E-04s)\n",
      "2021-08-05 11:54:45.889 - amici.ode_export - INFO - Finished computing dsigmaydp                +++ (1.25E-02s)\n",
      "2021-08-05 11:54:45.893 - amici.ode_export - INFO - Finished writing dsigmaydp.cpp               ++ (1.97E-02s)\n",
      "2021-08-05 11:54:45.900 - amici.ode_export - INFO - Finished writing sigmay.cpp                  ++ (1.18E-03s)\n",
      "2021-08-05 11:54:45.909 - amici.ode_export - INFO - Finished computing stau                     +++ (1.31E-04s)\n",
      "2021-08-05 11:54:45.910 - amici.ode_export - INFO - Finished writing stau.cpp                    ++ (4.64E-03s)\n",
      "2021-08-05 11:54:45.920 - amici.ode_export - INFO - Finished computing deltax                   +++ (1.20E-04s)\n",
      "2021-08-05 11:54:45.921 - amici.ode_export - INFO - Finished writing deltax.cpp                  ++ (5.07E-03s)\n",
      "2021-08-05 11:54:45.932 - amici.ode_export - INFO - Finished computing deltasx                  +++ (1.17E-04s)\n",
      "2021-08-05 11:54:45.933 - amici.ode_export - INFO - Finished writing deltasx.cpp                 ++ (4.24E-03s)\n",
      "2021-08-05 11:54:46.170 - amici.ode_export - INFO - Finished writing w.cpp                       ++ (2.30E-01s)\n",
      "2021-08-05 11:54:46.185 - amici.ode_export - INFO - Finished simplifying x0                    ++++ (3.60E-03s)\n",
      "2021-08-05 11:54:46.186 - amici.ode_export - INFO - Finished computing x0                       +++ (7.28E-03s)\n",
      "2021-08-05 11:54:46.191 - amici.ode_export - INFO - Finished writing x0.cpp                      ++ (1.57E-02s)\n",
      "2021-08-05 11:54:46.203 - amici.ode_export - INFO - Finished simplifying x0_fixedParameters    ++++ (5.97E-04s)\n",
      "2021-08-05 11:54:46.204 - amici.ode_export - INFO - Finished computing x0_fixedParameters       +++ (5.32E-03s)\n",
      "2021-08-05 11:54:46.207 - amici.ode_export - INFO - Finished writing x0_fixedParameters.cpp      ++ (1.06E-02s)\n",
      "2021-08-05 11:54:46.318 - amici.ode_export - INFO - Finished running smart_jacobian            ++++ (9.74E-02s)\n",
      "2021-08-05 11:54:46.334 - amici.ode_export - INFO - Finished simplifying sx0                   ++++ (9.60E-03s)\n",
      "2021-08-05 11:54:46.335 - amici.ode_export - INFO - Finished computing sx0                      +++ (1.18E-01s)\n",
      "2021-08-05 11:54:46.342 - amici.ode_export - INFO - Finished writing sx0.cpp                     ++ (1.28E-01s)\n",
      "2021-08-05 11:54:46.354 - amici.ode_export - INFO - Finished running smart_jacobian            ++++ (4.76E-04s)\n",
      "2021-08-05 11:54:46.359 - amici.ode_export - INFO - Finished running smart_jacobian            ++++ (1.72E-04s)\n",
      "2021-08-05 11:54:46.364 - amici.ode_export - INFO - Finished simplifying sx0_fixedParameters   ++++ (1.63E-04s)\n",
      "2021-08-05 11:54:46.365 - amici.ode_export - INFO - Finished computing sx0_fixedParameters      +++ (1.54E-02s)\n",
      "2021-08-05 11:54:46.367 - amici.ode_export - INFO - Finished writing sx0_fixedParameters.cpp     ++ (2.01E-02s)\n",
      "2021-08-05 11:54:46.681 - amici.ode_export - INFO - Finished writing xdot.cpp                    ++ (3.07E-01s)\n",
      "2021-08-05 11:54:46.687 - amici.ode_export - INFO - Finished writing y.cpp                       ++ (1.39E-03s)\n",
      "2021-08-05 11:54:46.699 - amici.ode_export - INFO - Finished simplifying x_rdata               ++++ (3.40E-04s)\n",
      "2021-08-05 11:54:46.700 - amici.ode_export - INFO - Finished computing x_rdata                  +++ (4.27E-03s)\n",
      "2021-08-05 11:54:46.721 - amici.ode_export - INFO - Finished writing x_rdata.cpp                 ++ (2.87E-02s)\n",
      "2021-08-05 11:54:46.735 - amici.ode_export - INFO - Finished simplifying total_cl              ++++ (6.01E-05s)\n",
      "2021-08-05 11:54:46.737 - amici.ode_export - INFO - Finished computing total_cl                 +++ (4.87E-03s)\n",
      "2021-08-05 11:54:46.740 - amici.ode_export - INFO - Finished writing total_cl.cpp                ++ (1.12E-02s)\n",
      "2021-08-05 11:54:46.754 - amici.ode_export - INFO - Finished simplifying x_solver              ++++ (4.87E-04s)\n",
      "2021-08-05 11:54:46.754 - amici.ode_export - INFO - Finished computing x_solver                 +++ (5.05E-03s)\n",
      "2021-08-05 11:54:46.780 - amici.ode_export - INFO - Finished writing x_solver.cpp                ++ (3.30E-02s)\n",
      "2021-08-05 11:54:46.847 - amici.ode_export - INFO - Finished generating cpp code                  + (2.50E+01s)\n",
      "2021-08-05 11:55:19.075 - amici.ode_export - INFO - Finished compiling cpp code                   + (3.22E+01s)\n",
      "2021-08-05 11:55:19.102 - amici.petab_import - INFO - Finished Importing PEtab model                (6.44E+01s)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "running build_ext\n",
      "Changed extra_compile_args for unix to ['-Xpreprocessor', '-fopenmp', '-std=c++14']\n",
      "building 'thermo_raf._thermo_raf' extension\n",
      "Testing SWIG executable swig4.0... FAILED.\n",
      "Testing SWIG executable swig3.0... FAILED.\n",
      "Testing SWIG executable swig... SUCCEEDED.\n",
      "swigging swig/thermo_raf.i to swig/thermo_raf_wrap.cpp\n",
      "swig -python -c++ -modern -outdir thermo_raf -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/swig -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/include -o swig/thermo_raf_wrap.cpp swig/thermo_raf.i\n",
      "Deprecated command line option: -modern. This option is now always on.\n",
      "creating build\n",
      "creating build/temp.macosx-10.14-x86_64-3.9\n",
      "creating build/temp.macosx-10.14-x86_64-3.9/swig\n",
      "clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/amici_models/thermo_raf -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/gsl -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/sundials/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/SuiteSparse/include -I/usr/local/include -I/usr/local/include -I/usr/local/opt/openssl@1.1/include -I/usr/local/opt/sqlite/include -I/usr/local/opt/tcl-tk/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/include -I/usr/local/Cellar/python@3.9/3.9.2_1/Frameworks/Python.framework/Versions/3.9/include/python3.9 -c swig/thermo_raf_wrap.cpp -o build/temp.macosx-10.14-x86_64-3.9/swig/thermo_raf_wrap.o -Xpreprocessor -fopenmp -std=c++14\n",
      "clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/amici_models/thermo_raf -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/gsl -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/sundials/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/SuiteSparse/include -I/usr/local/include -I/usr/local/include -I/usr/local/opt/openssl@1.1/include -I/usr/local/opt/sqlite/include -I/usr/local/opt/tcl-tk/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/include -I/usr/local/Cellar/python@3.9/3.9.2_1/Frameworks/Python.framework/Versions/3.9/include/python3.9 -c thermo_raf.cpp -o build/temp.macosx-10.14-x86_64-3.9/thermo_raf.o -Xpreprocessor -fopenmp -std=c++14\n",
      "clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/amici_models/thermo_raf -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/gsl -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/sundials/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/SuiteSparse/include -I/usr/local/include -I/usr/local/include -I/usr/local/opt/openssl@1.1/include -I/usr/local/opt/sqlite/include -I/usr/local/opt/tcl-tk/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/include -I/usr/local/Cellar/python@3.9/3.9.2_1/Frameworks/Python.framework/Versions/3.9/include/python3.9 -c thermo_raf_Jy.cpp -o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_Jy.o -Xpreprocessor -fopenmp -std=c++14\n",
      "clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/amici_models/thermo_raf -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/gsl -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/sundials/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/SuiteSparse/include -I/usr/local/include -I/usr/local/include -I/usr/local/opt/openssl@1.1/include -I/usr/local/opt/sqlite/include -I/usr/local/opt/tcl-tk/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/include -I/usr/local/Cellar/python@3.9/3.9.2_1/Frameworks/Python.framework/Versions/3.9/include/python3.9 -c thermo_raf_dJydsigma.cpp -o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_dJydsigma.o -Xpreprocessor -fopenmp -std=c++14\n",
      "clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/amici_models/thermo_raf -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/gsl -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/sundials/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/SuiteSparse/include -I/usr/local/include -I/usr/local/include -I/usr/local/opt/openssl@1.1/include -I/usr/local/opt/sqlite/include -I/usr/local/opt/tcl-tk/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/include -I/usr/local/Cellar/python@3.9/3.9.2_1/Frameworks/Python.framework/Versions/3.9/include/python3.9 -c thermo_raf_dJydy.cpp -o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_dJydy.o -Xpreprocessor -fopenmp -std=c++14\n",
      "clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/amici_models/thermo_raf -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/gsl -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/sundials/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/SuiteSparse/include -I/usr/local/include -I/usr/local/include -I/usr/local/opt/openssl@1.1/include -I/usr/local/opt/sqlite/include -I/usr/local/opt/tcl-tk/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/include -I/usr/local/Cellar/python@3.9/3.9.2_1/Frameworks/Python.framework/Versions/3.9/include/python3.9 -c thermo_raf_dJydy_colptrs.cpp -o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_dJydy_colptrs.o -Xpreprocessor -fopenmp -std=c++14\n",
      "clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/amici_models/thermo_raf -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/gsl -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/sundials/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/SuiteSparse/include -I/usr/local/include -I/usr/local/include -I/usr/local/opt/openssl@1.1/include -I/usr/local/opt/sqlite/include -I/usr/local/opt/tcl-tk/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/include -I/usr/local/Cellar/python@3.9/3.9.2_1/Frameworks/Python.framework/Versions/3.9/include/python3.9 -c thermo_raf_dJydy_rowvals.cpp -o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_dJydy_rowvals.o -Xpreprocessor -fopenmp -std=c++14\n",
      "clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/amici_models/thermo_raf -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/gsl -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/sundials/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/SuiteSparse/include -I/usr/local/include -I/usr/local/include -I/usr/local/opt/openssl@1.1/include -I/usr/local/opt/sqlite/include -I/usr/local/opt/tcl-tk/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/include -I/usr/local/Cellar/python@3.9/3.9.2_1/Frameworks/Python.framework/Versions/3.9/include/python3.9 -c thermo_raf_dsigmaydp.cpp -o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_dsigmaydp.o -Xpreprocessor -fopenmp -std=c++14\n",
      "clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/amici_models/thermo_raf -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/gsl -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/sundials/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/SuiteSparse/include -I/usr/local/include -I/usr/local/include -I/usr/local/opt/openssl@1.1/include -I/usr/local/opt/sqlite/include -I/usr/local/opt/tcl-tk/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/include -I/usr/local/Cellar/python@3.9/3.9.2_1/Frameworks/Python.framework/Versions/3.9/include/python3.9 -c thermo_raf_dwdp.cpp -o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_dwdp.o -Xpreprocessor -fopenmp -std=c++14\n",
      "clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/amici_models/thermo_raf -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/gsl -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/sundials/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/SuiteSparse/include -I/usr/local/include -I/usr/local/include -I/usr/local/opt/openssl@1.1/include -I/usr/local/opt/sqlite/include -I/usr/local/opt/tcl-tk/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/include -I/usr/local/Cellar/python@3.9/3.9.2_1/Frameworks/Python.framework/Versions/3.9/include/python3.9 -c thermo_raf_dwdp_colptrs.cpp -o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_dwdp_colptrs.o -Xpreprocessor -fopenmp -std=c++14\n",
      "clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/amici_models/thermo_raf -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/gsl -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/sundials/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/SuiteSparse/include -I/usr/local/include -I/usr/local/include -I/usr/local/opt/openssl@1.1/include -I/usr/local/opt/sqlite/include -I/usr/local/opt/tcl-tk/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/include -I/usr/local/Cellar/python@3.9/3.9.2_1/Frameworks/Python.framework/Versions/3.9/include/python3.9 -c thermo_raf_dwdp_rowvals.cpp -o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_dwdp_rowvals.o -Xpreprocessor -fopenmp -std=c++14\n",
      "clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/amici_models/thermo_raf -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/gsl -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/sundials/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/SuiteSparse/include -I/usr/local/include -I/usr/local/include -I/usr/local/opt/openssl@1.1/include -I/usr/local/opt/sqlite/include -I/usr/local/opt/tcl-tk/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/include -I/usr/local/Cellar/python@3.9/3.9.2_1/Frameworks/Python.framework/Versions/3.9/include/python3.9 -c thermo_raf_dwdw.cpp -o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_dwdw.o -Xpreprocessor -fopenmp -std=c++14\n",
      "clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/amici_models/thermo_raf -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/gsl -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/sundials/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/SuiteSparse/include -I/usr/local/include -I/usr/local/include -I/usr/local/opt/openssl@1.1/include -I/usr/local/opt/sqlite/include -I/usr/local/opt/tcl-tk/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/include -I/usr/local/Cellar/python@3.9/3.9.2_1/Frameworks/Python.framework/Versions/3.9/include/python3.9 -c thermo_raf_dwdw_colptrs.cpp -o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_dwdw_colptrs.o -Xpreprocessor -fopenmp -std=c++14\n",
      "clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/amici_models/thermo_raf -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/gsl -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/sundials/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/SuiteSparse/include -I/usr/local/include -I/usr/local/include -I/usr/local/opt/openssl@1.1/include -I/usr/local/opt/sqlite/include -I/usr/local/opt/tcl-tk/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/include -I/usr/local/Cellar/python@3.9/3.9.2_1/Frameworks/Python.framework/Versions/3.9/include/python3.9 -c thermo_raf_dwdw_rowvals.cpp -o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_dwdw_rowvals.o -Xpreprocessor -fopenmp -std=c++14\n",
      "clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/amici_models/thermo_raf -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/gsl -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/sundials/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/SuiteSparse/include -I/usr/local/include -I/usr/local/include -I/usr/local/opt/openssl@1.1/include -I/usr/local/opt/sqlite/include -I/usr/local/opt/tcl-tk/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/include -I/usr/local/Cellar/python@3.9/3.9.2_1/Frameworks/Python.framework/Versions/3.9/include/python3.9 -c thermo_raf_dwdx.cpp -o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_dwdx.o -Xpreprocessor -fopenmp -std=c++14\n",
      "clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/amici_models/thermo_raf -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/gsl -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/sundials/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/SuiteSparse/include -I/usr/local/include -I/usr/local/include -I/usr/local/opt/openssl@1.1/include -I/usr/local/opt/sqlite/include -I/usr/local/opt/tcl-tk/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/include -I/usr/local/Cellar/python@3.9/3.9.2_1/Frameworks/Python.framework/Versions/3.9/include/python3.9 -c thermo_raf_dwdx_colptrs.cpp -o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_dwdx_colptrs.o -Xpreprocessor -fopenmp -std=c++14\n",
      "clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/amici_models/thermo_raf -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/gsl -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/sundials/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/SuiteSparse/include -I/usr/local/include -I/usr/local/include -I/usr/local/opt/openssl@1.1/include -I/usr/local/opt/sqlite/include -I/usr/local/opt/tcl-tk/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/include -I/usr/local/Cellar/python@3.9/3.9.2_1/Frameworks/Python.framework/Versions/3.9/include/python3.9 -c thermo_raf_dwdx_rowvals.cpp -o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_dwdx_rowvals.o -Xpreprocessor -fopenmp -std=c++14\n",
      "clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/amici_models/thermo_raf -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/gsl -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/sundials/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/SuiteSparse/include -I/usr/local/include -I/usr/local/include -I/usr/local/opt/openssl@1.1/include -I/usr/local/opt/sqlite/include -I/usr/local/opt/tcl-tk/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/include -I/usr/local/Cellar/python@3.9/3.9.2_1/Frameworks/Python.framework/Versions/3.9/include/python3.9 -c thermo_raf_dxdotdw.cpp -o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_dxdotdw.o -Xpreprocessor -fopenmp -std=c++14\n",
      "clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/amici_models/thermo_raf -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/gsl -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/sundials/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/SuiteSparse/include -I/usr/local/include -I/usr/local/include -I/usr/local/opt/openssl@1.1/include -I/usr/local/opt/sqlite/include -I/usr/local/opt/tcl-tk/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/include -I/usr/local/Cellar/python@3.9/3.9.2_1/Frameworks/Python.framework/Versions/3.9/include/python3.9 -c thermo_raf_dxdotdw_colptrs.cpp -o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_dxdotdw_colptrs.o -Xpreprocessor -fopenmp -std=c++14\n",
      "clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/amici_models/thermo_raf -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/gsl -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/sundials/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/SuiteSparse/include -I/usr/local/include -I/usr/local/include -I/usr/local/opt/openssl@1.1/include -I/usr/local/opt/sqlite/include -I/usr/local/opt/tcl-tk/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/include -I/usr/local/Cellar/python@3.9/3.9.2_1/Frameworks/Python.framework/Versions/3.9/include/python3.9 -c thermo_raf_dxdotdw_rowvals.cpp -o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_dxdotdw_rowvals.o -Xpreprocessor -fopenmp -std=c++14\n",
      "clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/amici_models/thermo_raf -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/gsl -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/sundials/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/SuiteSparse/include -I/usr/local/include -I/usr/local/include -I/usr/local/opt/openssl@1.1/include -I/usr/local/opt/sqlite/include -I/usr/local/opt/tcl-tk/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/include -I/usr/local/Cellar/python@3.9/3.9.2_1/Frameworks/Python.framework/Versions/3.9/include/python3.9 -c thermo_raf_dydp.cpp -o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_dydp.o -Xpreprocessor -fopenmp -std=c++14\n",
      "clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/amici_models/thermo_raf -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/gsl -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/sundials/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/SuiteSparse/include -I/usr/local/include -I/usr/local/include -I/usr/local/opt/openssl@1.1/include -I/usr/local/opt/sqlite/include -I/usr/local/opt/tcl-tk/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/include -I/usr/local/Cellar/python@3.9/3.9.2_1/Frameworks/Python.framework/Versions/3.9/include/python3.9 -c thermo_raf_dydx.cpp -o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_dydx.o -Xpreprocessor -fopenmp -std=c++14\n",
      "clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/amici_models/thermo_raf -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/gsl -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/sundials/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/SuiteSparse/include -I/usr/local/include -I/usr/local/include -I/usr/local/opt/openssl@1.1/include -I/usr/local/opt/sqlite/include -I/usr/local/opt/tcl-tk/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/include -I/usr/local/Cellar/python@3.9/3.9.2_1/Frameworks/Python.framework/Versions/3.9/include/python3.9 -c thermo_raf_sigmay.cpp -o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_sigmay.o -Xpreprocessor -fopenmp -std=c++14\n",
      "clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/amici_models/thermo_raf -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/gsl -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/sundials/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/SuiteSparse/include -I/usr/local/include -I/usr/local/include -I/usr/local/opt/openssl@1.1/include -I/usr/local/opt/sqlite/include -I/usr/local/opt/tcl-tk/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/include -I/usr/local/Cellar/python@3.9/3.9.2_1/Frameworks/Python.framework/Versions/3.9/include/python3.9 -c thermo_raf_sx0.cpp -o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_sx0.o -Xpreprocessor -fopenmp -std=c++14\n",
      "clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/amici_models/thermo_raf -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/gsl -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/sundials/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/SuiteSparse/include -I/usr/local/include -I/usr/local/include -I/usr/local/opt/openssl@1.1/include -I/usr/local/opt/sqlite/include -I/usr/local/opt/tcl-tk/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/include -I/usr/local/Cellar/python@3.9/3.9.2_1/Frameworks/Python.framework/Versions/3.9/include/python3.9 -c thermo_raf_sx0_fixedParameters.cpp -o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_sx0_fixedParameters.o -Xpreprocessor -fopenmp -std=c++14\n",
      "clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/amici_models/thermo_raf -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/gsl -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/sundials/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/SuiteSparse/include -I/usr/local/include -I/usr/local/include -I/usr/local/opt/openssl@1.1/include -I/usr/local/opt/sqlite/include -I/usr/local/opt/tcl-tk/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/include -I/usr/local/Cellar/python@3.9/3.9.2_1/Frameworks/Python.framework/Versions/3.9/include/python3.9 -c thermo_raf_total_cl.cpp -o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_total_cl.o -Xpreprocessor -fopenmp -std=c++14\n",
      "clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/amici_models/thermo_raf -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/gsl -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/sundials/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/SuiteSparse/include -I/usr/local/include -I/usr/local/include -I/usr/local/opt/openssl@1.1/include -I/usr/local/opt/sqlite/include -I/usr/local/opt/tcl-tk/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/include -I/usr/local/Cellar/python@3.9/3.9.2_1/Frameworks/Python.framework/Versions/3.9/include/python3.9 -c thermo_raf_w.cpp -o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_w.o -Xpreprocessor -fopenmp -std=c++14\n",
      "clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/amici_models/thermo_raf -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/gsl -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/sundials/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/SuiteSparse/include -I/usr/local/include -I/usr/local/include -I/usr/local/opt/openssl@1.1/include -I/usr/local/opt/sqlite/include -I/usr/local/opt/tcl-tk/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/include -I/usr/local/Cellar/python@3.9/3.9.2_1/Frameworks/Python.framework/Versions/3.9/include/python3.9 -c thermo_raf_x0.cpp -o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_x0.o -Xpreprocessor -fopenmp -std=c++14\n",
      "clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/amici_models/thermo_raf -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/gsl -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/sundials/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/SuiteSparse/include -I/usr/local/include -I/usr/local/include -I/usr/local/opt/openssl@1.1/include -I/usr/local/opt/sqlite/include -I/usr/local/opt/tcl-tk/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/include -I/usr/local/Cellar/python@3.9/3.9.2_1/Frameworks/Python.framework/Versions/3.9/include/python3.9 -c thermo_raf_x0_fixedParameters.cpp -o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_x0_fixedParameters.o -Xpreprocessor -fopenmp -std=c++14\n",
      "clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/amici_models/thermo_raf -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/gsl -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/sundials/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/SuiteSparse/include -I/usr/local/include -I/usr/local/include -I/usr/local/opt/openssl@1.1/include -I/usr/local/opt/sqlite/include -I/usr/local/opt/tcl-tk/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/include -I/usr/local/Cellar/python@3.9/3.9.2_1/Frameworks/Python.framework/Versions/3.9/include/python3.9 -c thermo_raf_x_rdata.cpp -o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_x_rdata.o -Xpreprocessor -fopenmp -std=c++14\n",
      "clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/amici_models/thermo_raf -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/gsl -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/sundials/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/SuiteSparse/include -I/usr/local/include -I/usr/local/include -I/usr/local/opt/openssl@1.1/include -I/usr/local/opt/sqlite/include -I/usr/local/opt/tcl-tk/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/include -I/usr/local/Cellar/python@3.9/3.9.2_1/Frameworks/Python.framework/Versions/3.9/include/python3.9 -c thermo_raf_x_solver.cpp -o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_x_solver.o -Xpreprocessor -fopenmp -std=c++14\n",
      "clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/amici_models/thermo_raf -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/gsl -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/sundials/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/SuiteSparse/include -I/usr/local/include -I/usr/local/include -I/usr/local/opt/openssl@1.1/include -I/usr/local/opt/sqlite/include -I/usr/local/opt/tcl-tk/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/include -I/usr/local/Cellar/python@3.9/3.9.2_1/Frameworks/Python.framework/Versions/3.9/include/python3.9 -c thermo_raf_xdot.cpp -o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_xdot.o -Xpreprocessor -fopenmp -std=c++14\n",
      "clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/amici_models/thermo_raf -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/gsl -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/sundials/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/SuiteSparse/include -I/usr/local/include -I/usr/local/include -I/usr/local/opt/openssl@1.1/include -I/usr/local/opt/sqlite/include -I/usr/local/opt/tcl-tk/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/include -I/usr/local/Cellar/python@3.9/3.9.2_1/Frameworks/Python.framework/Versions/3.9/include/python3.9 -c thermo_raf_y.cpp -o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_y.o -Xpreprocessor -fopenmp -std=c++14\n",
      "clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/amici_models/thermo_raf -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/gsl -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/sundials/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/ThirdParty/SuiteSparse/include -I/usr/local/include -I/usr/local/include -I/usr/local/opt/openssl@1.1/include -I/usr/local/opt/sqlite/include -I/usr/local/opt/tcl-tk/include -I/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/include -I/usr/local/Cellar/python@3.9/3.9.2_1/Frameworks/Python.framework/Versions/3.9/include/python3.9 -c wrapfunctions.cpp -o build/temp.macosx-10.14-x86_64-3.9/wrapfunctions.o -Xpreprocessor -fopenmp -std=c++14\n",
      "clang++ -bundle -undefined dynamic_lookup build/temp.macosx-10.14-x86_64-3.9/swig/thermo_raf_wrap.o build/temp.macosx-10.14-x86_64-3.9/thermo_raf.o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_Jy.o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_dJydsigma.o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_dJydy.o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_dJydy_colptrs.o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_dJydy_rowvals.o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_dsigmaydp.o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_dwdp.o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_dwdp_colptrs.o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_dwdp_rowvals.o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_dwdw.o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_dwdw_colptrs.o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_dwdw_rowvals.o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_dwdx.o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_dwdx_colptrs.o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_dwdx_rowvals.o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_dxdotdw.o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_dxdotdw_colptrs.o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_dxdotdw_rowvals.o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_dydp.o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_dydx.o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_sigmay.o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_sx0.o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_sx0_fixedParameters.o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_total_cl.o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_w.o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_x0.o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_x0_fixedParameters.o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_x_rdata.o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_x_solver.o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_xdot.o build/temp.macosx-10.14-x86_64-3.9/thermo_raf_y.o build/temp.macosx-10.14-x86_64-3.9/wrapfunctions.o -L/usr/local/lib -L/Users/ffroehlich/Documents/HMS/energy_modeling_chapter/venv/lib/python3.9/site-packages/amici/libs -L/usr/local/lib -L/usr/local/opt/openssl@1.1/lib -L/usr/local/opt/sqlite/lib -L/usr/local/opt/tcl-tk/lib -lamici -lsundials -lsuitesparse -lcblas -lhdf5_hl_cpp -lhdf5_hl -lhdf5_cpp -lhdf5 -o /Users/ffroehlich/Documents/HMS/energy_modeling_chapter/amici_models/thermo_raf/thermo_raf/_thermo_raf.cpython-39-darwin.so -Xpreprocessor -fopenmp -lomp\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import pypesto\n",
    "import pypesto.petab\n",
    "importer = pypesto.petab.PetabImporter(petab_problem,\n",
    "                                       model_name=model.name)\n",
    "\n",
    "pypesto_problem = importer.create_problem(force_compile=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Simulations in AMICI are highly customizable and while many steps, including everything that was specified in the the petab problem, are automated by pypesto, not all options are set automatically and some user input may be required. For example, the considered example includes preequilibrations which is automatically handled by pyPESTO and AMICI. However, AMICI supports multiple different methods to compute steadystates as well as respective sensitivities. By default, a combination of Newton's method and simulation is employed to find the steadystate and sensitivites are computed using the Implicit Function Theorem (IFT), which is the most efficient option when applicable. Sensitivity computation via the IFT requires that the Jacobian of the model equations is not singular, which is not the case for models with conservation laws. As we do not account for protein synthesis and degradation, total protein abundances are conserved quantities in the model and the IFT is not applicable. Thus, we here employ a purely simulation based steadystate search by setting the allowed newton steps to 0 and compute the steadystate sensitivities using forward sensitivity analysis by setting the steadystate sensitivity mode accordingly."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "import amici\n",
    "pypesto_problem.objective.amici_solver.setNewtonMaxSteps(0)\n",
    "pypesto_problem.objective.amici_model.setSteadyStateSensitivityMode(amici.SteadyStateSensitivityMode.simulationFSA)\n",
    "pypesto_problem.objective.guess_steadystate = False"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "AMICI also supports different sensitivity analysis methods for the simulation itself. For models with over 100 parameters, it usually makes sense to use adjoint sensitivity analysis for simulation conditions. However, for models that also require preequilibration, the computation time of the simulation and gradient evaluation is dominated by the preequilibration part and the choice of sensitivity method for the simulation part will not have a significant impact on computation time. Accordingly, other considerations, such as the availability of approximate second order information should be taken into account when selecting the sensitivity method. In contrast to forward sensitivity analysis, adjoint sensitivity analysis does not provide an approximation to the Hessian of the objective function, rendering some optimizers inapplicable and potentially reducing the convergence rate of those that remain applicable. Therefore, we here chose forward sensitivty analysis for the simulation part."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "pypesto_problem.objective.amici_solver.setSensitivityMethod(amici.SensitivityMethod.forward)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Another set of crucial parameters to tune are integration step limits and tolerances. Lower tolerances generally improve accuracy of the objective function and its derivatives, which influences local convergence rate and success. However, lower tolerances also increases the computation time, so fewer optimization iterations could easily be offsetted by longer compute times per iteration. Moreover, both integration tolerances and step limits influence numerical integrability, which can lead to complications during optimization. For some optimizers, low step limits can prevent optimization from getting stuck in difficult to integrate parameter regions, but if parameters from these regions yield good fits to the data, this can also prevent optimization from reaching good fits. Accordingly, setting adequate values for step limits and tolerances is integral for model calibration. Nevertheless, there are little to no theoretical or practical guidelines on how to choose adequate values. Ranges of $10^{-8}$ to $10^{-12}$ are usually adequate, but most of the time values are chosen and updated empirically. Here we chose tolerances of $10^{-12}$ and a step limit of $10^{5}$ integration steps, which is rather conservative, but ensures adequate local convergence."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "pypesto_problem.objective.amici_solver.setAbsoluteTolerance(1e-12)\n",
    "pypesto_problem.objective.amici_solver.setRelativeTolerance(1e-12)\n",
    "pypesto_problem.objective.amici_solver.setMaxSteps(int(1e5))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Even with tuned integration tolerances and step limits, integration failures which prevent evaluation of the objective function and it's derivatives is usually inevitable. While some optimizers are able to continue after integration failures by rejecting the respective parameter update, it is impossible to recover from an integration failure during initialization. To still ensure that a sufficient number number of optimization runs is performed, pyPESTO enables resampling of startpoints, which generates new startpoints until the objective function could be successfully evaluated at all starting points. In the following, we activate this behaviour by passing an options object with the respective setting activated."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pypesto.optimize import OptimizeOptions\n",
    "optimize_options = OptimizeOptions(startpoint_resample=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Selecting an appropriate number of optimization runs can be challenging but should generally be adapted to the downstream analysis. For example, when model selection or uncertainty analysis using profile likelihoods will be performed, it is crucial to ensure that optimization reaches a local optimum and that the local optimum is, or at least yields a fit equivalent to, the global optimum. This can be achieved by running many optimization runs and allocating large computational budgets for each run. For large models, the total computational budget usually is insufficient to achieve these goals and the number of runs and the computational budget has to be limited, with an unknown tradeoff between the two. To deal with this tradeoff, we will here use a two-staged approach where we first run many starts with a low computational budget of 10 minutes and will then select the most promising runs and continue optimization with a longer computational budget. For this purpose, we use the fides optimizer, a high performance trust region optimizer which permits the direct specification of the computational budgets."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pypesto.optimize import FidesOptimizer\n",
    "import fides\n",
    "\n",
    "optimizer = FidesOptimizer(options={fides.Options.MAXTIME: 600})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we have everything at hand to run the first calibration stage, where run 100 optimization runs with uniformly sampled starting points. The calibration is parallelized using pyPESTOs multithreading engine with 4 threads."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Warning] AMICI:CVODES:CVode:WARNING: AMICI ERROR: in module CVODES in function CVode : Internal t = 351.988 and h = 2.22949e-14 are such that t + h = t on the next step. The solver will continue anyway. \n",
      "[Warning] AMICI:CVODES:CVode:WARNING: AMICI ERROR: in module CVODES in function CVode : Internal t = 351.988 and h = 2.22949e-14 are such that t + h = t on the next step. The solver will continue anyway. \n",
      "[Warning] AMICI:CVODES:CVode:WARNING: AMICI ERROR: in module CVODES in function CVode : Internal t = 351.988 and h = 2.22949e-14 are such that t + h = t on the next step. The solver will continue anyway. \n",
      "[Warning] AMICI:CVODES:CVode:WARNING: AMICI ERROR: in module CVODES in function CVode : Internal t = 351.988 and h = 2.22949e-14 are such that t + h = t on the next step. The solver will continue anyway. \n",
      "[Warning] AMICI:CVODES:CVode:WARNING: AMICI ERROR: in module CVODES in function CVode : Internal t = 351.988 and h = 2.22949e-14 are such that t + h = t on the next step. The solver will continue anyway. \n",
      "[Warning] AMICI:CVODES:CVode:WARNING: AMICI ERROR: in module CVODES in function CVode : Internal t = 351.988 and h = 2.22949e-14 are such that t + h = t on the next step. The solver will continue anyway. \n",
      "[Warning] AMICI:CVODES:CVode:WARNING: AMICI ERROR: in module CVODES in function CVode : Internal t = 351.988 and h = 2.22949e-14 are such that t + h = t on the next step. The solver will continue anyway. \n",
      "[Warning] AMICI:CVODES:CVode:WARNING: AMICI ERROR: in module CVODES in function CVode : Internal t = 351.988 and h = 2.22949e-14 are such that t + h = t on the next step. The solver will continue anyway. \n",
      "[Warning] AMICI:CVODES:CVode:WARNING: AMICI ERROR: in module CVODES in function CVode : Internal t = 351.988 and h = 2.22949e-14 are such that t + h = t on the next step. The solver will continue anyway. \n",
      "[Warning] AMICI:CVODES:CVode:WARNING: AMICI ERROR: in module CVODES in function CVode : Internal t = 351.988 and h = 2.22949e-14 are such that t + h = t on the next step. The solver will continue anyway. \n",
      "[Warning] AMICI:CVODES:CVode:WARNING: AMICI ERROR: in module CVODES in function CVode : The above warning has been issued mxhnil times and will not be issued again for this problem. \n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 8/96 in fxdot!\n",
      "[Warning] AMICI:CVODES:CVode:OTHER: AMICI ERROR: in module CVODES in function CVode : At t = 351.988 repeated recoverable right-hand side function errors. \n",
      "[Warning] AMICI:simulation: AMICI simulation failed:\n",
      "Steady state computation failed. First run of Newton solver failed: No convergence was achieved. Simulation to steady state failed. Second run of Newton solver failed: No convergence was achieved.\n",
      "Error occurred in:\n",
      "0          0x1350ae383 amici::SteadystateProblem::handleSteadyStateFailure(amici::Solver const*, amici::Model*) + 531\n",
      "1          0x1350ad65c amici::SteadystateProblem::findSteadyState(amici::Solver*, amici::NewtonSolver*, amici::Model*, int) + 332\n",
      "2          0x1350ad3be amici::SteadystateProblem::workSteadyStateProblem(amici::Solver*, amici::Model*, int) + 254\n",
      "3          0x13508eeb1 amici::AmiciApplication::runAmiciSimulation(amici::Solver&, amici::ExpData const*, amici::Model&, bool) + 241\n",
      "4  \n",
      "[Warning] AMICI:CVODES:CVode:WARNING: AMICI ERROR: in module CVODES in function CVode : Internal t = 0.0394337 and h = 2.72107e-18 are such that t + h = t on the next step. The solver will continue anyway. \n",
      "[Warning] AMICI:CVODES:CVode:WARNING: AMICI ERROR: in module CVODES in function CVode : Internal t = 0.0394337 and h = 2.72107e-18 are such that t + h = t on the next step. The solver will continue anyway. \n",
      "[Warning] AMICI:CVODES:CVode:WARNING: AMICI ERROR: in module CVODES in function CVode : Internal t = 0.0394337 and h = 2.72107e-18 are such that t + h = t on the next step. The solver will continue anyway. \n",
      "[Warning] AMICI:CVODES:CVode:WARNING: AMICI ERROR: in module CVODES in function CVode : Internal t = 0.0394337 and h = 2.72107e-18 are such that t + h = t on the next step. The solver will continue anyway. \n",
      "[Warning] AMICI:CVODES:CVode:WARNING: AMICI ERROR: in module CVODES in function CVode : Internal t = 0.0394337 and h = 2.72107e-18 are such that t + h = t on the next step. The solver will continue anyway. \n",
      "[Warning] AMICI:CVODES:CVode:WARNING: AMICI ERROR: in module CVODES in function CVode : Internal t = 0.0394337 and h = 2.72107e-18 are such that t + h = t on the next step. The solver will continue anyway. \n",
      "[Warning] AMICI:CVODES:CVode:WARNING: AMICI ERROR: in module CVODES in function CVode : Internal t = 0.0394337 and h = 2.72107e-18 are such that t + h = t on the next step. The solver will continue anyway. \n",
      "[Warning] AMICI:CVODES:CVode:WARNING: AMICI ERROR: in module CVODES in function CVode : Internal t = 0.0394337 and h = 2.72107e-18 are such that t + h = t on the next step. The solver will continue anyway. \n",
      "[Warning] AMICI:CVODES:CVode:WARNING: AMICI ERROR: in module CVODES in function CVode : Internal t = 0.0394337 and h = 2.72107e-18 are such that t + h = t on the next step. The solver will continue anyway. \n",
      "[Warning] AMICI:CVODES:CVode:WARNING: AMICI ERROR: in module CVODES in function CVode : Internal t = 0.0394337 and h = 2.72107e-18 are such that t + h = t on the next step. The solver will continue anyway. \n",
      "[Warning] AMICI:CVODES:CVode:WARNING: AMICI ERROR: in module CVODES in function CVode : The above warning has been issued mxhnil times and will not be issued again for this problem. \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:Inf: AMICI encountered an Inf value at index 38/96 in fxdot!\n",
      "[Warning] AMICI:CVODES:CVode:OTHER: AMICI ERROR: in module CVODES in function CVode : At t = 0.0394337 repeated recoverable right-hand side function errors. \n",
      "[Warning] AMICI:simulation: AMICI forward simulation failed at t = 0.039434:\n",
      "AMICI failed to integrate the forward problem\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Performing parallel task execution on 4 threads.\n",
      "  0%|                                                                                                                              | 0/100 [00:00<?, ?it/s]Executing task 0.\n",
      "100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 100/100 [00:00<00:00, 11064.43it/s]Executing task 1.\n",
      "\n",
      "Executing task 2.\n",
      "Executing task 3.\n",
      "[Warning] AMICI:simulation: AMICI simulation failed:\n",
      "Steady state computation failed. First run of Newton solver failed: No convergence was achieved. Simulation to steady state failed: No convergence was achieved. Second run of Newton solver failed: No convergence was achieved.\n",
      "Error occurred in:\n",
      "0          0x1350ae383 amici::SteadystateProblem::handleSteadyStateFailure(amici::Solver const*, amici::Model*) + 531\n",
      "1          0x1350ad65c amici::SteadystateProblem::findSteadyState(amici::Solver*, amici::NewtonSolver*, amici::Model*, int) + 332\n",
      "2          0x1350ad3be amici::SteadystateProblem::workSteadyStateProblem(amici::Solver*, amici::Model*, int) + 254\n",
      "3          0x13508eeb1 amici::AmiciApplication::runAmiciSimulation(amici::Solver&, amici::ExpData const*, amici::Model&, bool) + 241\n",
      "4  \n",
      "2021-08-05 11:59:15,414 - fides - INFO -   iter |    fval    | fval diff | pred diff | tr ratio  |  delta   |  ||g||   | ||step|| |  theta   |  alpha   | step | refl | trun | accept\n",
      "2021-08-05 11:59:15,415 - fides - INFO -      0 | +INF |    NaN    |    NaN    |    NaN    | 1.00E+00 | NAN |   NaN    |   NaN    |   NaN    | NaN  | NaN  | NaN  | 0\n",
      "Final fval=inf, time=55.2398s, n_fval=1.\n",
      "Executing task 4.\n",
      "2021-08-05 11:59:19,055 - fides - INFO -   iter |    fval    | fval diff | pred diff | tr ratio  |  delta   |  ||g||   | ||step|| |  theta   |  alpha   | step | refl | trun | accept\n",
      "2021-08-05 11:59:19,056 - fides - INFO -      0 | +4.253E+03 |    NaN    |    NaN    |    NaN    | 1.00E+00 | 3.26E+03 |   NaN    |   NaN    |   NaN    | NaN  | NaN  | NaN  | 1\n",
      "[Warning] AMICI:simulation: AMICI simulation failed:\n",
      "Steady state sensitivity computation failed due to unsuccessful factorization of RHS Jacobian\n",
      "Error occurred in:\n",
      "0          0x1350ad4b1 amici::SteadystateProblem::workSteadyStateProblem(amici::Solver*, amici::Model*, int) + 497\n",
      "1          0x13508eeb1 amici::AmiciApplication::runAmiciSimulation(amici::Solver&, amici::ExpData const*, amici::Model&, bool) + 241\n",
      "2          0x13508fd48 .omp_outlined. + 488\n",
      "3          0x1353663e3 __kmp_invoke_microtask + 147\n",
      "4          0x13532b7c7 __kmp_fork_call + 4310\n",
      "5          0x135321604 __kmpc_fork_call + 301\n",
      "6          0x13508f850 amici::runAmiciSimulations(amici::Solver\n",
      "2021-08-05 11:59:46,867 - fides - INFO -   iter |    fval    | fval diff | pred diff | tr ratio  |  delta   |  ||g||   | ||step|| |  theta   |  alpha   | step | refl | trun | accept\n",
      "2021-08-05 11:59:46,868 - fides - INFO -      0 | +INF |    NaN    |    NaN    |    NaN    | 1.00E+00 | NAN |   NaN    |   NaN    |   NaN    | NaN  | NaN  | NaN  | 0\n",
      "Final fval=inf, time=86.6804s, n_fval=1.\n",
      "Executing task 5.\n",
      "2021-08-05 12:00:46,183 - fides - INFO -   iter |    fval    | fval diff | pred diff | tr ratio  |  delta   |  ||g||   | ||step|| |  theta   |  alpha   | step | refl | trun | accept\n",
      "2021-08-05 12:00:46,184 - fides - INFO -      0 | +6.855E+03 |    NaN    |    NaN    |    NaN    | 1.00E+00 | 4.00E+03 |   NaN    |   NaN    |   NaN    | NaN  | NaN  | NaN  | 1\n",
      "[Warning] AMICI:simulation: AMICI simulation failed:\n",
      "Steady state computation failed. First run of Newton solver failed: No convergence was achieved. Simulation to steady state failed: No convergence was achieved. Second run of Newton solver failed: No convergence was achieved.\n",
      "Error occurred in:\n",
      "0          0x1350ae383 amici::SteadystateProblem::handleSteadyStateFailure(amici::Solver const*, amici::Model*) + 531\n",
      "1          0x1350ad65c amici::SteadystateProblem::findSteadyState(amici::Solver*, amici::NewtonSolver*, amici::Model*, int) + 332\n",
      "2          0x1350ad3be amici::SteadystateProblem::workSteadyStateProblem(amici::Solver*, amici::Model*, int) + 254\n",
      "3          0x13508eeb1 amici::AmiciApplication::runAmiciSimulation(amici::Solver&, amici::ExpData const*, amici::Model&, bool) + 241\n",
      "4  \n",
      "2021-08-05 12:00:48,337 - fides - INFO -   iter |    fval    | fval diff | pred diff | tr ratio  |  delta   |  ||g||   | ||step|| |  theta   |  alpha   | step | refl | trun | accept\n",
      "2021-08-05 12:00:48,339 - fides - INFO -      0 | +INF |    NaN    |    NaN    |    NaN    | 1.00E+00 | NAN |   NaN    |   NaN    |   NaN    | NaN  | NaN  | NaN  | 0\n",
      "Final fval=inf, time=92.9201s, n_fval=1.\n",
      "Executing task 6.\n",
      "2021-08-05 12:02:01,669 - fides - INFO -      1 | +6.855E+03 | +1.04E+05 | -2.75E+03 | -5.91E+01 | 1.00E+00 | 4.00E+03 | 3.27E+00 | 9.50E-01 | 1.00E+00 | nd   |    0 |    0 | 0\n",
      "2021-08-05 12:02:16,157 - fides - INFO -      1 | +4.253E+03 | +1.02E+06 | -1.24E+03 | -1.92E+03 | 1.00E+00 | 3.26E+03 | 2.39E+00 | 9.50E-01 | 1.00E+00 | nd   |    0 |    0 | 0\n",
      "2021-08-05 12:03:13,146 - fides - INFO -   iter |    fval    | fval diff | pred diff | tr ratio  |  delta   |  ||g||   | ||step|| |  theta   |  alpha   | step | refl | trun | accept\n",
      "2021-08-05 12:03:13,148 - fides - INFO -      0 | +6.913E+03 |    NaN    |    NaN    |    NaN    | 1.00E+00 | 3.77E+03 |   NaN    |   NaN    |   NaN    | NaN  | NaN  | NaN  | 1\n"
     ]
    }
   ],
   "source": [
    "from pypesto.optimize import minimize\n",
    "result = minimize(\n",
    "    pypesto_problem, optimizer, n_starts=100, options=optimize_options,\n",
    "    startpoint_method=pypesto.startpoint.uniform,\n",
    "    engine=pypesto.engine.MultiThreadEngine(4)\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To assess whether the quality of the calibration, we will now use petabs visualization feature. The visualization instructions are again specified in table, where, in the case of line plots, every row corresponds to a single pair of connected lines for data an simulation. Here we make use of the of the `petab.DATASET_ID` values that we specified in the measurements table. In the following we will generate seperate plots for each observable and each inhibitor, where every plot shows the dose response in the EGF stimulated and unstimulated condition."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "visualization_table = pd.DataFrame([\n",
    "    {\n",
    "        petab.PLOT_ID: f'{rafi}_{obs}',\n",
    "        petab.PLOT_TYPE_SIMULATION: petab.LINE_PLOT,\n",
    "        petab.PLOT_TYPE_DATA: petab.MEAN_AND_SD,\n",
    "        petab.DATASET_ID: condition,\n",
    "        petab.X_VALUES: 'RAFi_0',\n",
    "        petab.Y_VALUES: obs,\n",
    "        petab.X_SCALE: petab.LOG10,\n",
    "        petab.X_LABEL: rafi,\n",
    "        petab.Y_LABEL: obs,\n",
    "        petab.LEGEND_ENTRY: condition.split('__')[0]\n",
    "    }\n",
    "    for rafi in RAFis\n",
    "    for obs in observable_table.index\n",
    "    for condition in measurement_table[petab.DATASET_ID].unique()\n",
    "    if condition.split('__')[1] == rafi\n",
    "])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "visualization_table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "While experimental data for visualization is already available in the measurements table, simulations of the calibrated model still have to be computed. To generate model simulations, we extract the pre-calibrated parameter values for the best fit of the first calibration round. As optimization results are sorted in increasing order of objective function values, this can be achieved by taking the result at the first index. Simulation results can be extracted using the pypesto objective and the amici function `amici.petab_objective.rdatas_to_simulation_df`. As the results contains the full parameter vector including parameter values that were not estimated, but the objective only expects estimated parameters, the parameters need to be subsetted according to the free indices. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x = pypesto_problem.get_reduced_vector(result.optimize_result.list[0]['x'],\n",
    "                                       pypesto_problem.x_free_indices)\n",
    "simulation = pypesto_problem.objective(x, return_dict=True)\n",
    "simulation_df = amici.petab_objective.rdatas_to_simulation_df(\n",
    "    simulation['rdatas'],\n",
    "    model=pypesto_problem.objective.amici_model,\n",
    "    measurement_df=measurement_table,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "simulation_df = simulation_df.sort_values([petab.OBSERVABLE_ID, petab.PREEQUILIBRATION_CONDITION_ID, \n",
    "                                           petab.SIMULATION_CONDITION_ID]).reset_index()\n",
    "measurement_table = measurement_table.sort_values([petab.OBSERVABLE_ID, petab.PREEQUILIBRATION_CONDITION_ID, \n",
    "                                                   petab.SIMULATION_CONDITION_ID]).reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can now visualize both data and simulation using `petab.visualize.plot_data_and_simulation`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import petab.visualize\n",
    "petab.visualize.plot_data_and_simulation(\n",
    "    exp_data=measurement_table,\n",
    "    exp_conditions=condition_table,\n",
    "    sim_data=simulation_df,\n",
    "    vis_spec=visualization_table\n",
    ");"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The visualization suggests, that model simulations take values that are similar to the data they still show qualitatively different behaviour. This suggests that the calibrated parameter values are an adequate first guess,  that this is an adequate starting point for further refinement.\n",
    "\n",
    "Before we proceed with the second calibration round, we can check the accuracy of the gradient at the pre-calibrated parameter values. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "pd.set_option('display.max_rows', 100)\n",
    "pd.set_option('display.max_columns', 100)\n",
    "pypesto_problem.objective.x_names = [pypesto_problem.objective.x_names[ix] for ix in pypesto_problem.x_free_indices]\n",
    "pypesto_problem.objective.check_grad_multi_eps(x, multi_eps=[1e-1, 1e-3, 1e-5],\n",
    "                                               verbosity=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The function computes forward (`fd_c`), backward (`fd_b`) and central (`fd_c`) finite differences and compares them to the sensitivity based gradient (`grad`), using the optimal step size `eps` among the provided values. The results show a low error in the finite difference approximation (`fd_err`) indicating accurate evaluation of the objective function. Moreover, the small absolute (`abs_err`) and relative difference (`rel_err`) between sensitivity based gradient and finite difference approximation indicates accurate evaluation of the gradient. The large magnitude of the gradient suggest that we did not yet reach a local optimum and that further optimization will likely improve agreement between data and simulation.\n",
    "\n",
    "We will now perfom this refinement by instantiating a new optimizer with a larger computational budget of one hour"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "optimizer_fine = FidesOptimizer(options={fides.Options.MAXTIME: 3600})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To initialize the second calibration round at the pre-calibrated parameter values, we extract the respective values from the results object and set them as guesses in the pypesto problem."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_refine = 10\n",
    "pypesto_problem.x_guesses_full = np.vstack(result.optimize_result.get_for_key('x')[:n_refine])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For startpoint generation, pypesto will first start optimization runs at the provided parameter guesses and only sample new startpoint if more optimization runs than guesses are requested. As for the first calibration run, we run the calibration using the `minimize` function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "result_refined = minimize(\n",
    "    pypesto_problem, optimizer_fine, n_starts=n_refine,\n",
    "    engine=pypesto.engine.MultiThreadEngine(4)\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As for the pre-calibration, we now visualize the results using `petab` and the previously constructed visualization table."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x = pypesto_problem.get_reduced_vector(result_refined.optimize_result.list[0]['x'],\n",
    "                                       pypesto_problem.x_free_indices)\n",
    "simulation = pypesto_problem.objective(x, return_dict=True)\n",
    "simulation_df = amici.petab_objective.rdatas_to_simulation_df(\n",
    "    simulation['rdatas'],\n",
    "    model=importer.create_model(),\n",
    "    measurement_df=measurement_table,\n",
    ")\n",
    "\n",
    "simulation_df = simulation_df.sort_values([petab.OBSERVABLE_ID, petab.PREEQUILIBRATION_CONDITION_ID, \n",
    "                                           petab.SIMULATION_CONDITION_ID]).reset_index()\n",
    "measurement_table = measurement_table.sort_values([petab.OBSERVABLE_ID, petab.PREEQUILIBRATION_CONDITION_ID, \n",
    "                                                   petab.SIMULATION_CONDITION_ID]).reset_index()\n",
    "\n",
    "petab.visualize.plot_data_and_simulation(\n",
    "    exp_data=measurement_table,\n",
    "    exp_conditions=condition_table,\n",
    "    sim_data=simulation_df,\n",
    "    vis_spec=visualization_table\n",
    ");"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "These plots now indicate good qualitative and quantitative agreement between experimental data and model simulations. As a last step of the analysis, we now compare the objective function values across both calibration rounds using a waterfall plot."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pypesto.visualize import waterfall\n",
    "waterfall(results=[result,result_refined],\n",
    "          legends=['pre-calibration', 'calibration']);"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The waterfall plot shows that the "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (ebng chapter)",
   "language": "python",
   "name": "python3_enbg"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.2"
  },
  "pycharm": {
   "stem_cell": {
    "cell_type": "raw",
    "metadata": {
     "collapsed": false
    },
    "source": []
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
